<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>List of genomic annotations — gtf_list • CroCoNet</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="List of genomic annotations — gtf_list"><meta name="description" content="List of genomic annotations per species as GRanges objects. The human annotation is the GTF of Hg38 GENCODE release 32 primary assembly, while the gorilla and cynomolgus macaque annotations were created by transferring the human annotation onto the gorGor6 and macFas6 genomes via the tool Liftoff (https://github.com/agshumate/Liftoff). Each annotation was subsetted for the 300 genes that feature in this example dataset."><meta property="og:description" content="List of genomic annotations per species as GRanges objects. The human annotation is the GTF of Hg38 GENCODE release 32 primary assembly, while the gorilla and cynomolgus macaque annotations were created by transferring the human annotation onto the gorGor6 and macFas6 genomes via the tool Liftoff (https://github.com/agshumate/Liftoff). Each annotation was subsetted for the 300 genes that feature in this example dataset."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">CroCoNet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/CroCoNet.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Hellmann-Lab/CroCoNet/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>List of genomic annotations</h1>
      <small class="dont-index">Source: <a href="https://github.com/Hellmann-Lab/CroCoNet/blob/main/R/data.R" class="external-link"><code>R/data.R</code></a></small>
      <div class="d-none name"><code>gtf_list.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>List of genomic annotations per species as GRanges objects. The human annotation is the GTF of Hg38 GENCODE release 32 primary assembly, while the gorilla and cynomolgus macaque annotations were created by transferring the human annotation onto the gorGor6 and macFas6 genomes via the tool Liftoff (https://github.com/agshumate/Liftoff). Each annotation was subsetted for the 300 genes that feature in this example dataset.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="va">gtf_list</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="format">Format<a class="anchor" aria-label="anchor" href="#format"></a></h2>
    <p>A named list of 3 GRanges objects. Each object has 8 columns.</p>
<p>Attributes:</p><dl><dt>seqnames</dt>
<dd><p>Chromosome or contig name.</p></dd>

<dt>start</dt>
<dd><p>Genomic start location.</p></dd>

<dt>end</dt>
<dd><p>Genomic end location.</p></dd>

<dt>width</dt>
<dd><p>Width of the feature in base pairs.</p></dd>

<dt>strand</dt>
<dd><p>Genomic strand ("+" or "-").</p></dd>

<dt>source</dt>
<dd><p>The prediction program or public database where the annotations came from.</p></dd>

<dt>type</dt>
<dd><p>Feature type (gene, transcript, exon, CDS, UTR, start_codon, stop_codon or Selenocysteine).</p></dd>

<dt>score</dt>
<dd><p>The degree of confidence in the feature's existence and coordinates.</p></dd>

<dt>phase</dt>
<dd><p>One of '0', '1' or '2'. '0' means that the first base of the feature is the first base of a codon, '1' that the second base is the first base of a codon, and so on.</p></dd>

<dt>gene_id</dt>
<dd><p>Unique identifier of the gene.</p></dd>

<dt>gene_name</dt>
<dd><p>Name of the gene.</p></dd>

<dt>transcript_id</dt>
<dd><p>Unique identifier of the transcript.</p></dd>

<dt>transcript_name</dt>
<dd><p>Name of the transcript.</p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="source">Source<a class="anchor" aria-label="anchor" href="#source"></a></h2>
    <p>&lt;ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/gencode.v32.primary_assembly.annotation.gtf.gz&gt;, &lt;https://hgdownload.soe.ucsc.edu/goldenPath/gorGor6/bigZips/gorGor6.fa.gz&gt;, &lt;https://ftp.ensembl.org/pub/release-109/fasta/macaca_fascicularis/dna/Macaca_fascicularis.Macaca_fascicularis_6.0.dna_sm.toplevel.fa.gz&gt;</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Anita Térmeg.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

