<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ CroCoNet</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">CroCoNet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/CroCoNet.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Hellmann-Lab/CroCoNet/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="POU5F1_target_conservation.html">POU5F1_target_conservation</a></code>

        </dt>
        <dd>Cross-species conservation measures of the target genes in the POU5F1 module</dd>
      </dl><dl><dt>

          <code><a href="addDirectionality.html">addDirectionality()</a></code>

        </dt>
        <dd>Add directionality of interactions</dd>
      </dl><dl><dt>

          <code><a href="applyPower.html">applyPower()</a></code>

        </dt>
        <dd>Apply soft-thresholding</dd>
      </dl><dl><dt>

          <code><a href="assignInitialModules.html">assignInitialModules()</a></code>

        </dt>
        <dd>Assign initial modules</dd>
      </dl><dl><dt>

          <code><a href="calculateEdgeDivergence.html">calculateEdgeDivergence()</a></code>

        </dt>
        <dd>Calculate the divergence of intramodular edges</dd>
      </dl><dl><dt>

          <code><a href="calculateEigengenes.html">calculateEigengenes()</a></code>

        </dt>
        <dd>Calculate module eigengenes</dd>
      </dl><dl><dt>

          <code><a href="calculatePresStats.html">calculatePresStats()</a></code>

        </dt>
        <dd>Calculate preservation statistics</dd>
      </dl><dl><dt>

          <code><a href="calculateTreeStats.html">calculateTreeStats()</a></code>

        </dt>
        <dd>Calculate tree statistics across all modules</dd>
      </dl><dl><dt>

          <code><a href="clone2species.html">clone2species</a></code>

        </dt>
        <dd>Clone-species conversion</dd>
      </dl><dl><dt>

          <code><a href="consensus_network.html">consensus_network</a></code>

        </dt>
        <dd>Consensus network</dd>
      </dl><dl><dt>

          <code><a href="convertPresToDist.html">convertPresToDist()</a></code>

        </dt>
        <dd>Convert preservation statistics to distances across all modules</dd>
      </dl><dl><dt>

          <code><a href="convertToDT.html">convertToDT()</a></code>

        </dt>
        <dd>Convert a list of igraphs to a list of data tables</dd>
      </dl><dl><dt>

          <code><a href="convertToGraph.html">convertToGraph()</a></code>

        </dt>
        <dd>Convert a list of data frames to a list of igraphs</dd>
      </dl><dl><dt>

          <code><a href="createConsensus.html">createConsensus()</a></code>

        </dt>
        <dd>Create consensus network</dd>
      </dl><dl><dt>

          <code><a href="createRandomModules.html">createRandomModules()</a></code>

        </dt>
        <dd>Create random modules</dd>
      </dl><dl><dt>

          <code><a href="dist.html">dist</a></code>

        </dt>
        <dd>Distance measures of the pruned modules</dd>
      </dl><dl><dt>

          <code><a href="dist_jk.html">dist_jk</a></code>

        </dt>
        <dd>Distance measures of the original and jackknifed pruned modules</dd>
      </dl><dl><dt>

          <code><a href="eigengenes.html">eigengenes</a></code>

        </dt>
        <dd>Eigengenes</dd>
      </dl><dl><dt>

          <code><a href="eigengenes_per_species.html">eigengenes_per_species</a></code>

        </dt>
        <dd>Eigengenes per species</dd>
      </dl><dl><dt>

          <code><a href="filterModuleTrees.html">filterModuleTrees()</a></code>

        </dt>
        <dd>Filter modules based on their tree representations</dd>
      </dl><dl><dt>

          <code><a href="findConservedDivergedModules.html">findConservedDivergedModules()</a></code>

        </dt>
        <dd>Find conserved and diverged modules</dd>
      </dl><dl><dt>

          <code><a href="findConservedDivergedTargets.html">findConservedDivergedTargets()</a></code>

        </dt>
        <dd>Find conserved and diverged target genes within a module</dd>
      </dl><dl><dt>

          <code><a href="fitTreeStatsLm.html">fitTreeStatsLm()</a></code>

        </dt>
        <dd>Fit (weighted) linear model between tree-based statistics</dd>
      </dl><dl><dt>

          <code><a href="genes.html">genes</a></code>

        </dt>
        <dd>All genes</dd>
      </dl><dl><dt>

          <code><a href="getRegulators.html">getRegulators()</a></code>

        </dt>
        <dd>Get transcriptional regulators</dd>
      </dl><dl><dt>

          <code><a href="getTreeDf.html">getTreeDf()</a></code>

        </dt>
        <dd>Convert a tree to a data frame</dd>
      </dl><dl><dt>

          <code><a href="gtf_list.html">gtf_list</a></code>

        </dt>
        <dd>List of genomic annotations</dd>
      </dl><dl><dt>

          <code><a href="image_TRs.html">image_TRs</a></code>

        </dt>
        <dd>IMAGE transcriptional regulators</dd>
      </dl><dl><dt>

          <code><a href="initial_modules.html">initial_modules</a></code>

        </dt>
        <dd>Initial modules</dd>
      </dl><dl><dt>

          <code><a href="jaspar_core_TRs.html">jaspar_core_TRs</a></code>

        </dt>
        <dd>JASPAR 2024 vertebrate core transcriptional regulators</dd>
      </dl><dl><dt>

          <code><a href="jaspar_unvalidated_TRs.html">jaspar_unvalidated_TRs</a></code>

        </dt>
        <dd>JASPAR 2024 unvalidated transcriptional regulators</dd>
      </dl><dl><dt>

          <code><a href="lm_overall.html">lm_overall</a></code>

        </dt>
        <dd>Linear model for the characterization of overall module conservation</dd>
      </dl><dl><dt>

          <code><a href="loadNetworks.html">loadNetworks()</a></code>

        </dt>
        <dd>Load networks as a list of igraph objects</dd>
      </dl><dl><dt>

          <code><a href="module_conservation_overall.html">module_conservation_overall</a></code>

        </dt>
        <dd>Cross-species conservation measures per module</dd>
      </dl><dl><dt>

          <code><a href="network_list.html">network_list</a></code>

        </dt>
        <dd>List of networks</dd>
      </dl><dl><dt>

          <code><a href="network_list_raw.html">network_list_raw</a></code>

        </dt>
        <dd>#' SCE object list of the primate neural differentiation dataset #' #' A subset of the primate neural differentiation scRNA-seq dataset as a list of SCE objects. The data was collected during the early neural differentiation of human, gorilla and cynomolgus macaque iPS cells with 3 human, 2 gorilla and 4 cynomolgus cell lines (clones). Using a directed differentiation protocol, cells were differentiated into neural progenitor cells (NPCs) over the course of 9 days, and scRNA-seq data was obtained at six time points (days 0, 1, 3, 5, 7 and 9) during this process. Each element of the list is the SCE object of a clone with 300 genes and 100 cells, containing the raw counts, log-normalized counts and randomized quantile residuals as well as the metadata. #' #' @format A named list of 9 SCE objects. Each SCE object has 300 rows, 100 columns, 9 metadata columns and 3 assays. #' #' Metadata columns: #' <dl> #' <dt>species</dt> <dd><p>Name of the species.</p></dd> #' <dt>clone</dt> <dd><p>Name of the clone/cell line.</p></dd> #' <dt>day</dt> <dd><p>The day when the cell was collected.</p></dd> #' <dt>n_UMIs</dt> <dd><p>Number of UMIs detected.</p></dd> #' <dt>n_genes</dt> <dd><p>Number of genes detected.</p></dd> #' <dt>perc_mito</dt> <dd><p>Percent of mitochondrial reads.</p></dd> #' <dt>sizeFactor</dt> <dd><p>Size factor for scaling normalization, calculated per clone using [scran::computeSumFactors] and [scran::quickCluster].</p></dd> #' <dt>pseudotime</dt> <dd><p>Pseudotime inferred by [SCORPIUS::infer_trajectory].</p></dd> #' <dt>cell_type</dt> <dd><p>Cell type labels predicted by [SingleR::classifySingleR] using the embryoid body dataset from Rhodes et al. 2022 as reference.</p></dd> #' </dl> #' Assays: #' <dl> #' <dt>counts</dt> <dd><p>Raw counts.</p></dd> #' <dt>logcounts</dt> <dd><p>Log-normalized counts created by calculating size factors per clone using [scran::computeSumFactors] and [scran::quickCluster], log-normalizing by [scuttle::logNormCounts].</p></dd> #' <dt>rand_quantile_res</dt> <dd><p>Randomized quantile residuals calculated by [transformGamPoi::transformGamPoi].</p></dd> #' </dl> "sce_list" List of raw networks</dd>
      </dl><dl><dt>

          <code><a href="normalizeEdgeWeights.html">normalizeEdgeWeights()</a></code>

        </dt>
        <dd>Normalize edge weights between 0 and 1</dd>
      </dl><dl><dt>

          <code><a href="plotConservedDivergedModules.html">plotConservedDivergedModules()</a></code>

        </dt>
        <dd>Plot module conservation</dd>
      </dl><dl><dt>

          <code><a href="plotConservedDivergedTargets.html">plotConservedDivergedTargets()</a></code>

        </dt>
        <dd>Plot target conservation</dd>
      </dl><dl><dt>

          <code><a href="plotDistMats.html">plotDistMats()</a></code>

        </dt>
        <dd>Plot the distance matrices of clones based on module connectivity patterns</dd>
      </dl><dl><dt>

          <code><a href="plotEigengeneHeatmap.html">plotEigengeneHeatmap()</a></code>

        </dt>
        <dd>Plot module eigengenes as a heatmap</dd>
      </dl><dl><dt>

          <code><a href="plotEigengenesAlongPseudotime.html">plotEigengenesAlongPseudotime()</a></code>

        </dt>
        <dd>Plot module eigengenes along a pseudotime trajectory</dd>
      </dl><dl><dt>

          <code><a href="plotEigengenesViolin.html">plotEigengenesViolin()</a></code>

        </dt>
        <dd>Plot eigengene distributions per cell type and species</dd>
      </dl><dl><dt>

          <code><a href="plotExprAlongPseudotime.html">plotExprAlongPseudotime()</a></code>

        </dt>
        <dd>Plot expression profiles along a pseudotime trajectory</dd>
      </dl><dl><dt>

          <code><a href="plotExprHeatmap.html">plotExprHeatmap()</a></code>

        </dt>
        <dd>Plot expression profiles as a heatmap</dd>
      </dl><dl><dt>

          <code><a href="plotExprViolin.html">plotExprViolin()</a></code>

        </dt>
        <dd>Plot expression distributions per cell type and species</dd>
      </dl><dl><dt>

          <code><a href="plotModuleSizeDistribution.html">plotModuleSizeDistribution()</a></code>

        </dt>
        <dd>Plot the distribution of module sizes</dd>
      </dl><dl><dt>

          <code><a href="plotNetworks.html">plotNetworks()</a></code>

        </dt>
        <dd>Plot the networks of modules</dd>
      </dl><dl><dt>

          <code><a href="plotPresStatDistributions.html">plotPresStatDistributions()</a></code>

        </dt>
        <dd>Plot the distributions of preservation statistics</dd>
      </dl><dl><dt>

          <code><a href="plotPresStats.html">plotPresStats()</a></code>

        </dt>
        <dd>Plot cross-species VS within-species preservation statistics per species pair</dd>
      </dl><dl><dt>

          <code><a href="plotSumEigengeneHeatmap.html">plotSumEigengeneHeatmap()</a></code>

        </dt>
        <dd>Plot summarized eigengenes as a heatmap</dd>
      </dl><dl><dt>

          <code><a href="plotSumEigengenesLine.html">plotSumEigengenesLine()</a></code>

        </dt>
        <dd>Plot summarized eigengenes as a line plot</dd>
      </dl><dl><dt>

          <code><a href="plotSumExprHeatmap.html">plotSumExprHeatmap()</a></code>

        </dt>
        <dd>Plot summarized expression profiles as a heatmap</dd>
      </dl><dl><dt>

          <code><a href="plotSumExprLine.html">plotSumExprLine()</a></code>

        </dt>
        <dd>Plot summarized expression profiles as a line plot</dd>
      </dl><dl><dt>

          <code><a href="plotTreeStatDistributions.html">plotTreeStatDistributions()</a></code>

        </dt>
        <dd>Plot the distributions of tree-based statistics</dd>
      </dl><dl><dt>

          <code><a href="plotTreeStats.html">plotTreeStats()</a></code>

        </dt>
        <dd>Plot tree-based statistics against each other</dd>
      </dl><dl><dt>

          <code><a href="plotTrees.html">plotTrees()</a></code>

        </dt>
        <dd>Plot the tree representations of modules</dd>
      </dl><dl><dt>

          <code><a href="pres_stats.html">pres_stats</a></code>

        </dt>
        <dd>Preservation statistics of the pruned modules</dd>
      </dl><dl><dt>

          <code><a href="pres_stats_jk.html">pres_stats_jk</a></code>

        </dt>
        <dd>Preservation statistics of the original and jackknifed pruned modules</dd>
      </dl><dl><dt>

          <code><a href="pruneModules.html">pruneModules()</a></code>

        </dt>
        <dd>Prune modules</dd>
      </dl><dl><dt>

          <code><a href="pruneModules_UIK_adj.html">pruneModules_UIK_adj()</a></code>

        </dt>
        <dd>Prune modules based on the regulator-target adjacencies using dynamic filtering</dd>
      </dl><dl><dt>

          <code><a href="pruneModules_UIK_adj_kIM.html">pruneModules_UIK_adj_kIM()</a></code>

        </dt>
        <dd>Prune modules based on the regulator-target adjacencies and intramodular connectivities using dynamic filtering</dd>
      </dl><dl><dt>

          <code><a href="pruneModules_topN.html">pruneModules_topN()</a></code>

        </dt>
        <dd>Prune modules based on the regulator-target adjacencies by keeping a fixed number of top targets</dd>
      </dl><dl><dt>

          <code><a href="pruned_modules.html">pruned_modules</a></code>

        </dt>
        <dd>Pruned modules</dd>
      </dl><dl><dt>

          <code><a href="random_dist_jk.html">random_dist_jk</a></code>

        </dt>
        <dd>Distance measures of the the original and jackknifed random modules</dd>
      </dl><dl><dt>

          <code><a href="random_modules.html">random_modules</a></code>

        </dt>
        <dd>Random modules</dd>
      </dl><dl><dt>

          <code><a href="random_pres_stats.html">random_pres_stats</a></code>

        </dt>
        <dd>Preservation statistics of the random modules</dd>
      </dl><dl><dt>

          <code><a href="random_pres_stats_jk.html">random_pres_stats_jk</a></code>

        </dt>
        <dd>Preservation statistics of the the original and jackknifed random modules</dd>
      </dl><dl><dt>

          <code><a href="random_tree_stats.html">random_tree_stats</a></code>

        </dt>
        <dd>Tree-based statistics of the random modules</dd>
      </dl><dl><dt>

          <code><a href="random_tree_stats_jk.html">random_tree_stats_jk</a></code>

        </dt>
        <dd>Tree-based statistics of the original and jackknifed random modules</dd>
      </dl><dl><dt>

          <code><a href="random_trees_jk.html">random_trees_jk</a></code>

        </dt>
        <dd>Trees of the original and jackknifed random modules</dd>
      </dl><dl><dt>

          <code><a href="reconstructTrees.html">reconstructTrees()</a></code>

        </dt>
        <dd>Reconstruct trees across all modules</dd>
      </dl><dl><dt>

          <code><a href="regulators.html">regulators</a></code>

        </dt>
        <dd>Transcriptional regulators</dd>
      </dl><dl><dt>

          <code><a href="removeOverlappingGenePairs.html">removeOverlappingGenePairs()</a></code>

        </dt>
        <dd>Remove gene pairs with overlapping annotations</dd>
      </dl><dl><dt>

          <code><a href="scaleFreeFitRSquared.html">scaleFreeFitRSquared()</a></code>

        </dt>
        <dd>Calculate \(R^2\) value of the scale-free model fit for the provided connectivities</dd>
      </dl><dl><dt>

          <code><a href="sce.html">sce</a></code>

        </dt>
        <dd>SCE object of the primate neural differentiation dataset</dd>
      </dl><dl><dt>

          <code><a href="selectPower.html">selectPower()</a></code>

        </dt>
        <dd>Select power for scale-free transfrmation</dd>
      </dl><dl><dt>

          <code><a href="summarizeJackknifeStats.html">summarizeJackknifeStats()</a></code>

        </dt>
        <dd>Summarize jackknife statistics</dd>
      </dl><dl><dt>

          <code><a href="summarizeStat.html">summarizeStat()</a></code>

        </dt>
        <dd>Summarize values of a statistic</dd>
      </dl><dl><dt>

          <code><a href="testPowers.html">testPowers()</a></code>

        </dt>
        <dd>Test soft-thresholding power values</dd>
      </dl><dl><dt>

          <code><a href="tree.html">tree</a></code>

        </dt>
        <dd>Phylogenetic tree</dd>
      </dl><dl><dt>

          <code><a href="tree_stats.html">tree_stats</a></code>

        </dt>
        <dd>Tree-based statistics of the pruned modules</dd>
      </dl><dl><dt>

          <code><a href="tree_stats_jk.html">tree_stats_jk</a></code>

        </dt>
        <dd>Tree-based statistics of the original and jackknifed pruned modules</dd>
      </dl><dl><dt>

          <code><a href="trees.html">trees</a></code>

        </dt>
        <dd>Trees of the pruned modules</dd>
      </dl><dl><dt>

          <code><a href="trees_jk.html">trees_jk</a></code>

        </dt>
        <dd>Trees of the original and jackknifed pruned modules</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Anita TÃ©rmeg.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

