<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Load networks as a list of igraph objects — loadNetworks • CroCoNet</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Load networks as a list of igraph objects — loadNetworks"><meta name="description" content="Loads the TSV files containing the gene-gene edge weights and summarizes them as a list of igraph objects per replicate (i.e. cell line or biological replicate within a species)."><meta property="og:description" content="Loads the TSV files containing the gene-gene edge weights and summarizes them as a list of igraph objects per replicate (i.e. cell line or biological replicate within a species)."><meta property="og:image" content="https://hellmann-lab.github.io/CroCoNet/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">CroCoNet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/CroCoNet.html">Tutorial</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Documentation</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Hellmann-Lab/CroCoNet/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Load networks as a list of igraph objects</h1>
      <small class="dont-index">Source: <a href="https://github.com/Hellmann-Lab/CroCoNet/blob/main/R/loadNetworks.R" class="external-link"><code>R/loadNetworks.R</code></a></small>
      <div class="d-none name"><code>loadNetworks.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Loads the TSV files containing the gene-gene edge weights and summarizes them as a list of <code>igraph</code> objects per replicate (i.e. cell line or biological replicate within a species).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">loadNetworks</span><span class="op">(</span></span>
<span>  <span class="va">path</span>,</span>
<span>  replicate_names <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rep <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  directed <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  min_occurrence <span class="op">=</span> <span class="fl">2L</span>,</span>
<span>  n_cores <span class="op">=</span> <span class="fl">1L</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-path">path<a class="anchor" aria-label="anchor" href="#arg-path"></a></dt>
<dd><p>Character, the path where the files containing the gene-gene edge weights can be found.</p></dd>


<dt id="arg-replicate-names">replicate_names<a class="anchor" aria-label="anchor" href="#arg-replicate-names"></a></dt>
<dd><p>Character vector, the names of the replicates that were used for network reconstruction. These are expected to be the base names of the TSV files (format: nameOfReplicate.tsv or nameOfReplicate_index.tsv). If set to NULL (default), the names are deduced from the file names by stripping anything starting with "_" and ".tsv".</p></dd>


<dt id="arg-rep">rep<a class="anchor" aria-label="anchor" href="#arg-rep"></a></dt>
<dd><p>Integer, the number of output files per replicate, e.g. the number of different subsamplings or the number of independent runs in case of a stochastic network inference algorithm (default: 1). If <code>rep</code> &gt; 1, the TSV files are expected to be indexed from 1 to <code>rep</code> for each replicate (format: nameOfReplicate_index.tsv).</p></dd>


<dt id="arg-directed">directed<a class="anchor" aria-label="anchor" href="#arg-directed"></a></dt>
<dd><p>Logical indicating whether the network inference method produces a directed output, i.e. whether geneA-geneB and geneB-geneA can both be present among the edges (default: TRUE). For GRNBoost2 this has to be left as TRUE, while for correlation-based methods this has to most likely be set to FALSE.</p></dd>


<dt id="arg-min-occurrence">min_occurrence<a class="anchor" aria-label="anchor" href="#arg-min-occurrence"></a></dt>
<dd><p>Integer, the minimum number of occurrences an edges has to have across runs/subsamplings to be kept (default: 2). Disregarded if <code>rep</code> = 1.</p></dd>


<dt id="arg-n-cores">n_cores<a class="anchor" aria-label="anchor" href="#arg-n-cores"></a></dt>
<dd><p>Integer, the number of cores (default: 1).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A named list of <code>igraph</code> objects containing the networks per replicate. The list names are taken from <code>replicate_names</code>. Each <code>igraph</code> object contains the edge attribute "weight" (taken from the 3rd column of the input TSV files). Starting from the 4th column of the input TSV files, all columns are preserved as edge attributes with unchanged names.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The functions reads and formats the TSV files in the directory specified by <code>path</code>. The base names of the TSV files of interest can be specified using the parameter <code>replicate_names</code> (in this case the files are expected to be named using the format: nameOfReplicate.tsv or nameOfReplicate_index.tsv). If  <code>replicate_names</code> is set to NULL (default), all TSV files in <code>path</code> are read and the replicate names are deduced from the file names by stripping anything starting with "_" and ".tsv".</p>
<p>As the first 3 columns, each input TSV file is expected to contain the first gene that forms the edge, the second gene that forms the edge and the edge weight. There can be additional columns as well, these will be preserved as edge attributes in the <code>igraph</code> objects.</p>
<p>If the edge weights were calculated on different subsamplings of cells per replicate, or a network inference algorithm involving stochastic steps (e.g. GRNBoost2) was run several times on each replicate, the edge weights are averaged across these subsamplings/runs, and a single combined <code>igraph</code> object is returned per replicate. The number of subsamplings/runs has to be specified by the parameter <code>rep</code>. The TSV files corresponding to the same replicate but different subsamplings/runs are expected to have the same base name with different indices: nameOfReplicate_index.tsv.</p>
<p>If the network inference method produces an output with directed edges, i.e. geneA-geneB and geneB-geneA can both be present, the parameter <code>directed</code> should be set to TRUE (this is the case e.g. for GRNBoost2). In this case the edge weights inferred between the same gene pair but in opposite directions are averaged. If the edge in one of the directions is missing, it is regarded as a 0. For correlation-based methods (e.g. <code>correlatePairs</code>), <code>directed</code> has to most likely be set to FALSE. In both cases, <code>loadNetworks</code> returns an undirected network.</p>
<p>Rarely occurring edges can be removed altogether by specifying <code>min_occurrence</code>. This is not relevant if the network inference was done only once per replicate (<code>rep</code> = 1), therefore in this case the value of <code>min_occurrence</code> is ignored. If the network inference was done several times for each replicate (<code>rep</code> &gt; 1), the highest possible number of occurrences for each edge is 2×<code>rep</code> in case of a directed network inference method and <code>rep</code> in case of an undirected network inference method. If an edge occurs less often than the specified value of <code>min_occurrence</code>, the edge is removed. This can be helpful to 1) denoise the networks and 2) decrease the computational power needed for the next steps.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Moerman, T., Santos, S. A., González-Blas, C. B., Simm, J., Moreau, Y., Aerts, J., &amp; Aerts, S. (2019). GRNBoost2 and Arboreto: efficient and scalable inference of gene regulatory networks. Bioinformatics , 35(12), 2159–2161.</p>
<p>Lun, A. T. L., McCarthy, D. J., &amp; Marioni, J. C. (2016). A step-by-step workflow for low-level analysis of single-cell RNA-seq data with Bioconductor. F1000Research, 5, 2122.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">network_list_raw</span> <span class="op">&lt;-</span> <span class="fu">loadNetworks</span><span class="op">(</span>path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"CroCoNet"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                                 rep <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Anita Térmeg.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

