<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Remove gene pairs with overlapping annotations — removeOverlappingGenePairs • CroCoNet</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Remove gene pairs with overlapping annotations — removeOverlappingGenePairs"><meta name="description" content="Removes the gene pairs (edges) that have overlapping annotations in the genomes of at least 1 species from all networks."><meta property="og:description" content="Removes the gene pairs (edges) that have overlapping annotations in the genomes of at least 1 species from all networks."><meta property="og:image" content="https://hellmann-lab.github.io/CroCoNet/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">CroCoNet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/CroCoNet.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Hellmann-Lab/CroCoNet/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Remove gene pairs with overlapping annotations</h1>
      <small class="dont-index">Source: <a href="https://github.com/Hellmann-Lab/CroCoNet/blob/main/R/removeOverlappingGenePairs.R" class="external-link"><code>R/removeOverlappingGenePairs.R</code></a></small>
      <div class="d-none name"><code>removeOverlappingGenePairs.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Removes the gene pairs (edges) that have overlapping annotations in the genomes of at least 1 species from all networks.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">removeOverlappingGenePairs</span><span class="op">(</span></span>
<span>  <span class="va">network_list</span>,</span>
<span>  <span class="va">gtf_list</span>,</span>
<span>  <span class="va">replicate2species</span>,</span>
<span>  gene_col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene_name"</span>, <span class="st">"gene_id"</span><span class="op">)</span>,</span>
<span>  n_cores <span class="op">=</span> <span class="fl">1L</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-network-list">network_list<a class="anchor" aria-label="anchor" href="#arg-network-list"></a></dt>
<dd><p>A named list of <code>igraph</code> objects containing the networks of all replicates.</p></dd>


<dt id="arg-gtf-list">gtf_list<a class="anchor" aria-label="anchor" href="#arg-gtf-list"></a></dt>
<dd><p>A named list of GRanges objects containing the genome annotations of all species.</p></dd>


<dt id="arg-replicate-species">replicate2species<a class="anchor" aria-label="anchor" href="#arg-replicate-species"></a></dt>
<dd><p>A data frame specifying which species each replicate belongs to, required columns:</p><dl><dt>replicate</dt>
<dd><p>Character, name of the replicate.</p></dd>

<dt>species</dt>
<dd><p>Character, name of the species.</p></dd>


</dl></dd>


<dt id="arg-gene-col">gene_col<a class="anchor" aria-label="anchor" href="#arg-gene-col"></a></dt>
<dd><p>Character specifying the type of identifier the network nodes have, one of "gene_name", "gene_id". The function looks for the names of the network nodes in the corresponding column of the GTF files.</p></dd>


<dt id="arg-n-cores">n_cores<a class="anchor" aria-label="anchor" href="#arg-n-cores"></a></dt>
<dd><p>Integer, the number of cores (default: 1).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A named list of <code>igraph</code> objects containing the networks of all replicates, after the removal of gene pairs with overlapping annotations. A new edge attribute is added to all <code>igraph</code> objects:</p><dl><dt>genomic_dist</dt>
<dd><p>Numeric, the genomic distance of the 2 genes that form the edge (<code>Inf</code> if the 2 genes are annotated on different chromosomes/contigs).</p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Mapping and counting is problematic for overlapping genomic features: it is difficult to tell apart which read belongs to which gene. Often parts of the reads from one gene are assigned to the other gene, leading to correlated expression profiles simply due to genomic position. This only has a marginal effect on the results of a DE analysis, but can cause false positive edges with very high edge weights in case of a network analysis.</p>
<p>This functions circumvents such potential artefacts by removing all edges between overlapping genes. As the first step, the positions of the network genes are determined in the genome of each species. Depending on the value of <code>gene_col</code>, the names of the network nodes are matched to the entries in either the "gene_name" or the "gene_id" column of the provided GTF files. As the second step, gene pairs with overlapping annotations are identified in each genome. Finally, these gene pairs are removed from all networks (not just from the networks of the species where they were found to overlap!).</p>
<p>For all remaining edges of the networks, the function adds the genomic distance between the 2 genes that form the edge as the edge attribute "genomic_dist". If the genes are annotated on different chromosomes, the distance is set to <code>Inf</code>. This information can be used for further sanity checking, e.g. to check the relationship between edge weight and genomic proximity.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">network_list_filt</span> <span class="op">&lt;-</span> <span class="fu">removeOverlappingGenePairs</span><span class="op">(</span><span class="va">network_list_raw</span>,</span></span>
<span class="r-in"><span>                                                <span class="va">gtf_list</span>,</span></span>
<span class="r-in"><span>                                                <span class="va">replicate2species</span>,</span></span>
<span class="r-in"><span>                                                <span class="st">"gene_name"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of gene pairs that overlap in any of the genomes:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 40480</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Anita Térmeg.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

