<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>SCE object of the primate neural differentiation dataset — sce • CroCoNet</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="SCE object of the primate neural differentiation dataset — sce"><meta name="description" content="A subset of the primate neural differentiation scRNA-seq dataset in an SCE format. The data was collected during the early neural differentiation of human, gorilla and cynomolgus macaque iPS cells with 3 human, 2 gorilla and 4 cynomolgus cell lines (replicates). Using a directed differentiation protocol, cells were differentiated into neural progenitor cells (NPCs) over the course of 9 days, and scRNA-seq data was obtained at six time points (days 0, 1, 3, 5, 7 and 9) during this process. The SCE object contains the raw and log-normalized counts as well as the metadata for 300 genes and 900 cells (100 cells per replicate)."><meta property="og:description" content="A subset of the primate neural differentiation scRNA-seq dataset in an SCE format. The data was collected during the early neural differentiation of human, gorilla and cynomolgus macaque iPS cells with 3 human, 2 gorilla and 4 cynomolgus cell lines (replicates). Using a directed differentiation protocol, cells were differentiated into neural progenitor cells (NPCs) over the course of 9 days, and scRNA-seq data was obtained at six time points (days 0, 1, 3, 5, 7 and 9) during this process. The SCE object contains the raw and log-normalized counts as well as the metadata for 300 genes and 900 cells (100 cells per replicate)."><meta property="og:image" content="https://hellmann-lab.github.io/CroCoNet/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">CroCoNet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/CroCoNet.html">Tutorial</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Documentation</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Hellmann-Lab/CroCoNet/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>SCE object of the primate neural differentiation dataset</h1>
      <small class="dont-index">Source: <a href="https://github.com/Hellmann-Lab/CroCoNet/blob/main/R/data.R" class="external-link"><code>R/data.R</code></a></small>
      <div class="d-none name"><code>sce.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>A subset of the primate neural differentiation scRNA-seq dataset in an SCE format. The data was collected during the early neural differentiation of human, gorilla and cynomolgus macaque iPS cells with 3 human, 2 gorilla and 4 cynomolgus cell lines (replicates). Using a directed differentiation protocol, cells were differentiated into neural progenitor cells (NPCs) over the course of 9 days, and scRNA-seq data was obtained at six time points (days 0, 1, 3, 5, 7 and 9) during this process. The SCE object contains the raw and log-normalized counts as well as the metadata for 300 genes and 900 cells (100 cells per replicate).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="va">sce</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="format">Format<a class="anchor" aria-label="anchor" href="#format"></a></h2>
    <p>An SCE object with 300 rows, 900 columns, 9 metadata columns and 2 assays.</p>
<p>Metadata columns:</p><dl><dt>species</dt>
<dd><p>Name of the species.</p></dd>

<dt>replicate</dt>
<dd><p>Name of the replicate/cell line.</p></dd>

<dt>day</dt>
<dd><p>The day when the cell was collected.</p></dd>

<dt>n_UMIs</dt>
<dd><p>Number of UMIs detected.</p></dd>

<dt>n_genes</dt>
<dd><p>Number of genes detected.</p></dd>

<dt>perc_mito</dt>
<dd><p>Percent of mitochondrial reads.</p></dd>

<dt>sizeFactor</dt>
<dd><p>Size factor for scaling normalization, calculated first per replicate using [scran::computeSumFactors] and [scran::quickCluster], then adjusted by [batchelor::multiBatchNorm] to remove systematic differences in covergae across replicates.</p></dd>

<dt>pseudotime</dt>
<dd><p>Pseudotime inferred by [SCORPIUS::infer_trajectory].</p></dd>

<dt>cell_type</dt>
<dd><p>Cell type labels predicted by [SingleR::classifySingleR] using the embryoid body dataset from Rhodes et al. 2022 as reference.</p></dd>


</dl><p>Assays:</p><dl><dt>counts</dt>
<dd><p>Raw counts.</p></dd>

<dt>logcounts</dt>
<dd><p>Log-normalized counts created by first calculating size factors per replicate using [scran::computeSumFactors] and [scran::quickCluster], then adjusted them to remove systematic differences in covergae across replicates and log-normalizing by [batchelor::multiBatchNorm].</p></dd>


</dl></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Anita Térmeg.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

