<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>CroCoNet • CroCoNet</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="CroCoNet">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">CroCoNet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/CroCoNet.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Hellmann-Lab/CroCoNet/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<script src="CroCoNet_files/kePrint-0.0.1/kePrint.js"></script><link href="CroCoNet_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>CroCoNet</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Hellmann-Lab/CroCoNet/blob/main/vignettes/CroCoNet.Rmd" class="external-link"><code>vignettes/CroCoNet.Rmd</code></a></small>
      <div class="d-none name"><code>CroCoNet.Rmd</code></div>
    </div>

    
    
<style>
img {
  border: 0;
}
p.caption {
  font-size: 0.95em;
  width: 85%;
  margin: auto;
}

<!-- h1.title { -->
<!--   font-size: 30px; -->
<!-- } -->
<!-- h1 { /* Header 1 */ -->
<!--   font-size: 24px; -->
<!-- } -->
<!-- h2 { /* Header 2 */ -->
<!--   font-size: 18px; -->
<!-- } -->

/* ---- Works in plain html_vignette ---- */
h1.title { font-size: 1.9rem; }   /* ≈ 30px (~22pt) */
h1        { font-size: 1.5rem; }  /* ≈ 24px (~18pt) */
h2        { font-size: 1.15rem; } /* ≈ 18px (~13pt) */

/* ---- Ensure the same sizes inside pkgdown ---- */
.template-article .page-header h1,
.template-vignette .page-header h1 {
  font-size: 1.9rem !important;
}

.template-article .contents h2,
.template-vignette .contents h2 {
  font-size: 1.5rem !important;
}

.template-article .contents h3,
.template-vignette .contents h3 {
  font-size: 1.15rem !important;
}

/* Optional: adjust spacing */
.template-article .contents h2,
.template-article .contents h3 {
  line-height: 1.25;
  margin-top: 1.1em;
}

body {
text-align: justify
}

</style>
<p><br><tt>CroCoNet</tt> (<strong>Cro</strong>ss-species
<strong>Co</strong>mparison of <strong>Net</strong>works) is a
computational pipeline and R package to quantitatively compare
gene-regulatory networks across species. The approach hinges on
contrasting network variability within and across species in order to
distinguish divergence from detection uncertainty. By enabling robust
network analyses, CroCoNet contributes towards finding meaningful
cross-species differences in gene regulation.</p>
<p>In this vignette, we demonstrate the complete CroCoNet workflow using
a single-cell RNA-seq dataset of early primate neural differentiation as
an example. The corresponding input data are available in the associated
Zenodo repository, under the compressed archive
“neural_differentiation_dataset.zip”. All file paths used in the
examples below are relative to this master folder.</p>
<p>Through this tutorial, users will learn how to:</p>
<ul>
<li><p>Prepare and format input data for CroCoNet (including
preprocessing and network inference)</p></li>
<li><p>Process networks and build the consensus network</p></li>
<li><p>Assign and characterize modules</p></li>
<li><p>Calculate module preservation scores within and across
species</p></li>
<li><p>Quantify module conservation across all species and along
specific lineages</p></li>
<li><p>Visualize modules using network- and tree-based representations
<br><br></p></li>
</ul>
<div class="section level2">
<h2 id="input-data">Input data<a class="anchor" aria-label="anchor" href="#input-data"></a>
</h2>
<div class="section level3">
<h3 id="experimental-design">Experimental design<a class="anchor" aria-label="anchor" href="#experimental-design"></a>
</h3>
<p><tt>CroCoNet</tt> was designed to analyze scRNA-seq data from
multiple species, with multiple replicates per species. For a meaningful
analysis, the conditions, pseudotemporal trajectory and cell type
composition should be as comparable across species as possible. To
reliably infer networks, the dataset should in addition have sufficient
expression variance across cells (several different conditions,
time-series experiment over a dynamic process or a mixture of cell
types).</p>
<p>To generate the example data, three iPS cell lines from three humans,
two iPS cell lines from one gorilla and four iPS cell lines from two
cynomolgus macaques were differentiated into neural progenitor cells
(NPCs) over the course of nine days. Samples from six different time
points were sequenced using scRNA-seq. The protocol and timeline were
the same for all species to make the results comparable.</p>
<p>At several steps of the CroCoNet workflow, the replicates need to be
matched to species. To make this possible, a data frame
<tt>replicate2species</tt> has to be created that specifies which
species each replicate belongs to. How this looks like for the example
data is shown below:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">replicate2species</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"processed_data/replicate2species.rds"</span><span class="op">)</span></span></code></pre></div>
<table class="table table" style="font-size: 13px; margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
replicate
</th>
<th style="text-align:left;">
species
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
H1c1
</td>
<td style="text-align:left;">
human
</td>
</tr>
<tr>
<td style="text-align:left;">
H2c1
</td>
<td style="text-align:left;">
human
</td>
</tr>
<tr>
<td style="text-align:left;">
H3c1
</td>
<td style="text-align:left;">
human
</td>
</tr>
<tr>
<td style="text-align:left;">
G1c1
</td>
<td style="text-align:left;">
gorilla
</td>
</tr>
<tr>
<td style="text-align:left;">
G1c2
</td>
<td style="text-align:left;">
gorilla
</td>
</tr>
<tr>
<td style="text-align:left;">
C1c1
</td>
<td style="text-align:left;">
cynomolgus
</td>
</tr>
<tr>
<td style="text-align:left;">
C1c2
</td>
<td style="text-align:left;">
cynomolgus
</td>
</tr>
<tr>
<td style="text-align:left;">
C2c1
</td>
<td style="text-align:left;">
cynomolgus
</td>
</tr>
<tr>
<td style="text-align:left;">
C2c2
</td>
<td style="text-align:left;">
cynomolgus
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="initial-processing">Initial processing<a class="anchor" aria-label="anchor" href="#initial-processing"></a>
</h3>
<p>The pipeline assumes that the user has already done the standard
processing steps on the transcriptomic data, including mapping, QC and
normalization. In addition, it is required to have a shared feature
space across all species. In case of closely related species, genes can
be matched by transferring the annotation of one species to the genomes
of others using the tool <em>Liftoff</em> <span class="citation">[1]</span>. It is of course also possible to use
orthology information or even protein language model-based macrogenes
<span class="citation">[2]</span>.</p>
<p>It is also recommended to perform cell type annotation and pseudotime
inference if applicable. After these steps, a SingleCellExperiment (SCE)
object should be created that contains the raw counts, normalized counts
and metadata (including information on species, replicate, cell type
assignment and inferred pseudotime).</p>
<p>In case of the primate differentiation dataset, we created a shared
feature space by transferring the human genome annotation to the gorGor6
and macFas6 genomes via <em>Liftoff</em>. We performed the nornalization
using <tt>scran</tt> <span class="citation">[3]</span>, pseudotime
inference using <tt>SCORPIUS</tt> <span class="citation">[4]</span> and
cell type annotation using <tt>singleR</tt> <span class="citation">[5]</span> with the embryoid body dataset from Rhodes
et al. <span class="citation">[6]</span> as the reference. We found that
the pseudotime trajectories are reasonably well-aligned across the 3
species, confirming comparability.</p>
<div class="figure" style="text-align: center">
<img src="trajectory.png" alt="Pseudotime trajectory of the early primate neural differentiation dataset coloured by sampling day, pseudotime, species and cell type." width="90%"><p class="caption">
Pseudotime trajectory of the early primate neural differentiation
dataset coloured by sampling day, pseudotime, species and cell type.
</p>
</div>
</div>
<div class="section level3">
<h3 id="network-inference">Network inference<a class="anchor" aria-label="anchor" href="#network-inference"></a>
</h3>
<p>Before applying the pipeline, the user must first infer co-expression
networks per replicate using a method of their choice, for example
<em>GRNBoost2</em> <span class="citation">[7]</span> (metric: gradient
boosting-based importance scores) or <em>correlatePairs</em> <span class="citation">[3]</span> (metric: Spearman’s correlation). The
networks should not be restricted to connections between transcriptional
regulators and their target genes, because <tt>CroCoNet</tt> relies also
on connections between the target genes themselves.</p>
<p>If the replicates differ substantially in cell type composition, it
is recommended to downsample them so that the proportions of cell types
are consistent across replicates. This helps disentangle
species-specific effects from cell type differences. It can be
beneficial to create multiple downsampled versions of each replicate and
run network inference on each, ensuring that valuable data are not
discarded.</p>
<p>If the chosen network inference algorithm involves stochastic steps
(as is the case for <em>GRNBoost2</em>), the results can be made more
robust by running the algorithm multiple times on the same data with
different random seeds.</p>
<p>The network reconstructions should be stored as TSV files where each
row corresponds to an edge. There should be at least 3 columns
containing 1) the 1st gene that forms the edge, 2) the 2nd gene that
forms the edge, and 3) the edge weight. The TSV files should follow the
naming convention <em>nameOfReplicate.tsv</em> (if there is a single
network per replicate) or <em>nameOfReplicate_index.tsv</em> (if
multiple network outputs are generated per replicate).</p>
<p>In case of example dataset, we inferred networks using GRNBoost2 with
all genes as potential regulators. We run the algorithm 10 times on the
count matrix of each replicate, producing a total of 9×10 TSV files
stored in the directory <tt>network_inference/output/</tt>. <br><br></p>
</div>
</div>
<div class="section level2">
<h2 id="loading-and-processing-the-networks">Loading and processing the networks<a class="anchor" aria-label="anchor" href="#loading-and-processing-the-networks"></a>
</h2>
<p>In the first part of the workflow, the output files of the network
inference are loaded as <tt>igraph</tt> objects, the edges are filtered
and the edge weights are normalized (if desired), and finally, a
phylogeny-aware consensus network is calculated across all replicates
and species. If the network inference algorithm does not distinguish
positively and negatively correlated gene pairs, this information can
also be added at this stage.</p>
<div class="section level3">
<h3 id="loading-the-networks-as-igraphs">Loading the networks as igraphs<a class="anchor" aria-label="anchor" href="#loading-the-networks-as-igraphs"></a>
</h3>
<p>As the first step, the TSV files containing the network
reconstructions are loaded and summarized as a list of <tt>igraph</tt>
<span class="citation">[8]</span> objects per replicate with the help of
the <tt>loadNetworks</tt> function.</p>
<p>If there are several versions of network reconstruction per replicate
due to different subsamplings or repeated runs (specified via the
parameter <tt>rep</tt>), the edge weights are averaged across these
versions, and a single combined <tt>igraph</tt> object is returned for
each replicate. If the network inference method produces an output with
directed edges, i.e. geneA-geneB and geneB-geneA can both be present
(specified via the parameter <tt>directed</tt>), the edge weights
inferred between the same gene pair but in opposite directions are also
averaged. Edges that fall below the detection threshold in a network
version/direction are assigned an edge weight of 0 for the calculation
of the mean. If an edge meets the detection criteria in too few
versions/directions (specified via the parameter
<tt>min_occurrence</tt>), it is assigned a final weight of 0 in the
corresponding replicate, regardless of its mean weight. This filtering
approach helps to 1) denoise the networks and 2) decrease the
computational power needed for the next steps. The steps relying on a
detection threshold are not applicable to correlation-based
networks.</p>
<p>In case of the example dataset, the 9×10 TSV files located in the
<tt>network_inference/output/</tt> directory were loaded as a list of 9
<tt>igraph</tt> objects (one network reconstruction for each of the 9
replicates). The parameter <tt>rep</tt> was set to 10 as GRNBoost2 had
been run 10 times on each replicate, and the parameter <tt>directed</tt>
was left at the default value of TRUE, as GRNBoost2 produces a directed
output. As a result, the edge weights were averaged across the 10 runs
and 2 directions for each edge and replicate. Since the parameter
<tt>min_occurrece</tt> was left at its default value of 2, edges that
occurred zero or one times across all runs and directions were assigned
a weight of 0.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">network_list_raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadNetworks.html">loadNetworks</a></span><span class="op">(</span><span class="st">"network_inference/output/"</span>, </span>
<span>                                 replicate_names <span class="op">=</span> <span class="va">replicate2species</span><span class="op">$</span><span class="va">replicate</span>,</span>
<span>                                 rep <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                                 directed <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                 min_occurrence <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p>The resulting igraph objects store the weights for each edge:</p>
<table class="table table" style="font-size: 13px; margin-left: auto; margin-right: auto;">
<caption style="font-size: initial !important;">
<span style="font-size:13px">A few example edges from the raw network of
the replicate C2c2</span>
</caption>
<thead><tr>
<th style="text-align:center;">
from
</th>
<th style="text-align:center;">
to
</th>
<th style="text-align:center;">
weight
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ACAT2
</td>
<td style="text-align:center;">
0.0041
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ACBD4
</td>
<td style="text-align:center;">
0.0107
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ACTR10
</td>
<td style="text-align:center;">
0.0187
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ACVR2A
</td>
<td style="text-align:center;">
0.0184
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ADAM9
</td>
<td style="text-align:center;">
0.0545
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ADARB1
</td>
<td style="text-align:center;">
0.0317
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ADM5
</td>
<td style="text-align:center;">
0.0404
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ADO
</td>
<td style="text-align:center;">
0.0180
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="removing-gene-pairs-with-overlapping-annotations">Removing gene pairs with overlapping annotations<a class="anchor" aria-label="anchor" href="#removing-gene-pairs-with-overlapping-annotations"></a>
</h3>
<p>Mapping and counting is problematic for overlapping genomic features:
parts of the reads from one gene can be assigned to the other gene,
leading to correlated expression profiles simply due to genomic
position. This has only a marginal effect on the results of a
differential expression analysis, but can cause false positive edges
with very high edge weights in case of a network analysis. Such
potential artefacts can be circumvented by removing all edges between
genes that have overlapping annotations in the genome(s) of at least one
species.</p>
<p>To do this, the genome annotations are needed in a list format:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gtf_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>human <span class="op">=</span> <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/io-gff-read.html" class="external-link">read_gff</a></span><span class="op">(</span><span class="st">"genomes/hg38.gtf"</span><span class="op">)</span>,</span>
<span>                 gorilla <span class="op">=</span> <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/io-gff-read.html" class="external-link">read_gff</a></span><span class="op">(</span><span class="st">"genomes/gorGor6.gtf"</span><span class="op">)</span>,</span>
<span>                 cynomolgus <span class="op">=</span> <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/io-gff-read.html" class="external-link">read_gff</a></span><span class="op">(</span><span class="st">"genomes/macFas6.gtf"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Then the function <tt>removeOverlappingGenePairs</tt> determines the
genomic positions of the network genes based on the GTFs, identifies
gene pairs with overlapping annotations in each genome and removes these
gene pairs from all networks:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">network_list_raw_filt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/removeOverlappingGenePairs.html">removeOverlappingGenePairs</a></span><span class="op">(</span><span class="va">network_list_raw</span>, </span>
<span>                                                    gtf_list <span class="op">=</span> <span class="va">gtf_list</span>, </span>
<span>                                                    replicate2species <span class="op">=</span> <span class="va">replicate2species</span>, </span>
<span>                                                    gene_col <span class="op">=</span> <span class="st">"gene_name"</span><span class="op">)</span></span></code></pre></div>
<p>During this step, the genomic distance is also added for each gene
pair of the network (the distance is regarded as Inf if the two genes
are located on different chromosomes):</p>
<table class="table table" style="font-size: 13px; margin-left: auto; margin-right: auto;">
<caption style="font-size: initial !important;">
<span style="font-size:13px">A few example edges from the network of the
replicate C2c2 after removing gene pairs with overlapping annotation and
adding genomic distances</span>
</caption>
<thead><tr>
<th style="text-align:center;">
from
</th>
<th style="text-align:center;">
to
</th>
<th style="text-align:center;">
weight
</th>
<th style="text-align:center;">
genomic_dist
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ACAT2
</td>
<td style="text-align:center;">
0.0041
</td>
<td style="text-align:center;">
Inf
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ACBD4
</td>
<td style="text-align:center;">
0.0107
</td>
<td style="text-align:center;">
Inf
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ACTR10
</td>
<td style="text-align:center;">
0.0187
</td>
<td style="text-align:center;">
Inf
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ACVR2A
</td>
<td style="text-align:center;">
0.0184
</td>
<td style="text-align:center;">
Inf
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ADAM9
</td>
<td style="text-align:center;">
0.0545
</td>
<td style="text-align:center;">
Inf
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ADARB1
</td>
<td style="text-align:center;">
0.0317
</td>
<td style="text-align:center;">
Inf
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ADM5
</td>
<td style="text-align:center;">
0.0404
</td>
<td style="text-align:center;">
8270693
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ADO
</td>
<td style="text-align:center;">
0.0180
</td>
<td style="text-align:center;">
Inf
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="normalizing-edge-weights">Normalizing edge weights<a class="anchor" aria-label="anchor" href="#normalizing-edge-weights"></a>
</h3>
<p>Normalizing the edge weights between 0 and 1 is recommended, because
it makes them interpretable as adjacencies and ensures that network
concepts such as connectivity are applicable.</p>
<p>There are 2 approaches for the normalization implemented as part of
the <tt>nornmalizeEdgeWeights</tt> function:</p>
<ul>
<li><p><strong>Unsigned network</strong> (default): Gene pairs with high
negative edge weights are considered as connected as gene pairs with
high positive edge weights. Therefore the negative edge weights are
first replaced by their absolute values, then all edge weights are
scaled by the maximum weight across all networks:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>a</mi><mo>=</mo><mfrac><mrow><mo stretchy="true" form="prefix">|</mo><mi>w</mi><mo stretchy="true" form="postfix">|</mo></mrow><mrow><mi>m</mi><mi>a</mi><mi>x</mi><mrow><mo stretchy="true" form="prefix">(</mo><mrow><mo stretchy="true" form="prefix">|</mo><mi>w</mi><mo stretchy="true" form="postfix">|</mo></mrow><mo stretchy="true" form="postfix">)</mo></mrow></mrow></mfrac></mrow><annotation encoding="application/x-tex">a =  \frac{|w|}{max(|w|)}</annotation></semantics></math>
where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>a</mi><annotation encoding="application/x-tex">a</annotation></semantics></math>
is the edge weight (adjacency) after normalization and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>w</mi><annotation encoding="application/x-tex">w</annotation></semantics></math>
is the edge weight before normalization. After the transformation, the
adjacencies around 0 correspond to the former low positive and low
negative values, while the adjacencies around 1 correspond to the former
high positive and high negative values.</p></li>
<li><p><strong>Signed network</strong>: Gene pairs with high negative
edge weights are considered unconnected. Therefore all edge weights are
transformed between 0 and 1 using a min-max normalization:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>a</mi><mo>=</mo><mfrac><mrow><mi>w</mi><mo>−</mo><mi>m</mi><mi>i</mi><mi>n</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>w</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><mrow><mi>m</mi><mi>a</mi><mi>x</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>w</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>−</mo><mi>m</mi><mi>i</mi><mi>n</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>w</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></mfrac></mrow><annotation encoding="application/x-tex">a = \frac{w - min(w)}{max(w) - min(w)}</annotation></semantics></math>
After the transformation, the adjacencies around 0 correspond to the
former high negative values and the adjacencies around 1 correspond to
the former high positive values.</p></li>
</ul>
<p>In case of the example dataset, we opted for an unsigned network:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">network_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalizeEdgeWeights.html">normalizeEdgeWeights</a></span><span class="op">(</span><span class="va">network_list_raw_filt</span>,</span>
<span>                                     signed <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<table class="table table" style="font-size: 13px; margin-left: auto; margin-right: auto;">
<caption style="font-size: initial !important;">
<span style="font-size:13px">A few example edges from the network of the
replicate C2c2 after normalizing edge weights</span>
</caption>
<thead><tr>
<th style="text-align:center;">
from
</th>
<th style="text-align:center;">
to
</th>
<th style="text-align:center;">
weight
</th>
<th style="text-align:center;">
genomic_dist
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ACAT2
</td>
<td style="text-align:center;">
0.000048
</td>
<td style="text-align:center;">
Inf
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ACBD4
</td>
<td style="text-align:center;">
0.000127
</td>
<td style="text-align:center;">
Inf
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ACTR10
</td>
<td style="text-align:center;">
0.000222
</td>
<td style="text-align:center;">
Inf
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ACVR2A
</td>
<td style="text-align:center;">
0.000218
</td>
<td style="text-align:center;">
Inf
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ADAM9
</td>
<td style="text-align:center;">
0.000647
</td>
<td style="text-align:center;">
Inf
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ADARB1
</td>
<td style="text-align:center;">
0.000376
</td>
<td style="text-align:center;">
Inf
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ADM5
</td>
<td style="text-align:center;">
0.000480
</td>
<td style="text-align:center;">
8270693
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ADO
</td>
<td style="text-align:center;">
0.000214
</td>
<td style="text-align:center;">
Inf
</td>
</tr>
</tbody>
</table>
<p>Using the resulting adjacencies, we also calculated the connectivity
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>k</mi><mi>i</mi></msub><annotation encoding="application/x-tex">k_i</annotation></semantics></math>)
of each gene, which is the sum of adjacencies between the given gene and
all others:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>k</mi><mi>i</mi></msub><mo>=</mo><munder><mo>∑</mo><mrow><mi>j</mi><mo>≠</mo><mi>i</mi></mrow></munder><msub><mi>a</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></mrow><annotation encoding="application/x-tex">k_i = \sum_{j \neq i} a_{ij}</annotation></semantics></math>
Both the adjacencies and connectivites have a highly left-skewed
distribution (the adjacencies range between 0 and 1, with a median of
0.000213, while the connectivities range between 0 and ~17, with a
median of 0.256). This is expected since connectivities in many
biological networks, including gene regulatory networks, have been
reported to follow scale-free distribution <span class="citation">[9–11]</span> (although the extent of this varies
across systems and metrics <span class="citation">[12]</span>).</p>
<div class="figure" style="text-align: center">
<img src="edge_weight_con_distr.png" alt="The distribution od adjacencies and connectivities for the networks of the 9 replicates." width="80%"><p class="caption">
The distribution od adjacencies and connectivities for the networks of
the 9 replicates.
</p>
</div>
</div>
<div class="section level3">
<h3 id="creating-the-consensus-network">Creating the consensus network<a class="anchor" aria-label="anchor" href="#creating-the-consensus-network"></a>
</h3>
<p>As the next step of the workflow, the networks across different
replicates and different species are integrated into a single consensus
network in a phylogeny-aware manner.</p>
<p>For each edge, the consensus adjacency is calculated as the weighted
mean of replicate-wise adjacencies. The weighted mean corrects for 1)
the phylogenetic distances between species, and 2) the different numbers
of replicates per species. As a result, the approach downweighs the edge
weights of the replicates that 1) belong to closely related species, or
2) belong to species with many replicates, so that an imbalanced
sampling across the phylogenetic tree does not bias the consensus
network.</p>
<p>If a phylogeny-aware consensus is desired, the phylogenetic tree has
to be provided. In case of the example dataset, this is the mammalian
tree from Bininda-Emonds et al. <span class="citation">[13]</span>
subsetted for Homo sapiens (human), Gorilla gorilla (gorilla) and Macaca
Fascicularis (cynomolgus):</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://static-content.springer.com/esm/art%3A10.1038%2Fnature05634/MediaObjects/41586_2007_BFnature05634_MOESM362_ESM.txt"</span>, <span class="fu">here</span><span class="op">(</span><span class="va">wd</span>, <span class="st">"mammalian_tree.txt"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">read.tree</span><span class="op">(</span><span class="fu">here</span><span class="op">(</span><span class="va">wd</span>, <span class="st">"mammalian_tree.txt"</span><span class="op">)</span><span class="op">)</span><span class="op">[[</span><span class="st">"treemammalST_bestDates="</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">keep.tip</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Homo_sapiens"</span>,<span class="st">"Gorilla_gorilla"</span>, <span class="st">"Macaca_fascicularis"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">tree</span><span class="op">$</span><span class="va">tip.label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cynomolgus"</span>, <span class="st">"gorilla"</span>, <span class="st">"human"</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="tree.png" alt="Phylogenetic tree of the species." width="28%"><p class="caption">
Phylogenetic tree of the species.
</p>
</div>
<p>Based on the phylogenetic tree and the species assignment of each
replicate, the consensus network is computed using the function
<tt>createConsensus</tt>:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">consensus_network</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createConsensus.html">createConsensus</a></span><span class="op">(</span><span class="va">network_list</span>, </span>
<span>                                     replicate2species <span class="op">=</span> <span class="va">replicate2species</span>, </span>
<span>                                     tree <span class="op">=</span> <span class="va">tree</span><span class="op">)</span></span></code></pre></div>
<table class="table table" style="font-size: 13px; margin-left: auto; margin-right: auto;">
<caption style="font-size: initial !important;">
<span style="font-size:13px">A few example edges from the consensus
network</span>
</caption>
<thead><tr>
<th style="text-align:center;">
from
</th>
<th style="text-align:center;">
to
</th>
<th style="text-align:center;">
weight
</th>
<th style="text-align:center;">
n_supporting_replicates
</th>
<th style="text-align:center;">
supporting_replicates
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ACAT2
</td>
<td style="text-align:center;">
0.0000426
</td>
<td style="text-align:center;">
3
</td>
<td style="text-align:center;">
H3c1,C2c1,C2c2
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ACBD4
</td>
<td style="text-align:center;">
0.0000145
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
C2c2
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ACBD6
</td>
<td style="text-align:center;">
0.0011186
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
C1c1
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ACBD7
</td>
<td style="text-align:center;">
0.0000081
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
C1c1
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ACER3
</td>
<td style="text-align:center;">
0.0000665
</td>
<td style="text-align:center;">
2
</td>
<td style="text-align:center;">
H1c1,H3c1
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ACIN1
</td>
<td style="text-align:center;">
0.0001691
</td>
<td style="text-align:center;">
2
</td>
<td style="text-align:center;">
G1c2,C1c1
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ACLY
</td>
<td style="text-align:center;">
0.0000402
</td>
<td style="text-align:center;">
2
</td>
<td style="text-align:center;">
H3c1,C2c1
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ACO1
</td>
<td style="text-align:center;">
0.0000244
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
C2c1
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="adding-the-direction-of-correlation">Adding the direction of correlation<a class="anchor" aria-label="anchor" href="#adding-the-direction-of-correlation"></a>
</h3>
<p>If the networks were inferred using a method that does not
distinguish between positively and negatively correlated gene pairs,
this information can be added for analyses where it is meaningful to
separate the two (e.g. when calculating eigengenes for activated VS
repressed targets of a transcriptional regulator). If the inference
method already outputs both positive and negative edge weights
(e.g. correlation-based methods), this information is already added
during the step “Normalizing edge weights” and does not have to be
calculated again.</p>
<p>If this step is required, the directionality of each edge is
determined by calculating a modified Spearman’s correlation between the
expression profiles of the two genes forming the edge and assigning the
sign of the modified Spearman’s rho as the direction. The calculation
uses the approximate Spearman’s rho, significance testing, and blocking
implemented in <tt>correlatePairs</tt> <span class="citation">[3]</span>.</p>
<p>In case of the example dataset, the networks were inferred using
GRNBoost2 which does not provide information about the direction of
regulation, therefore it is necessary to perform this step. To get the
expression profiles of the network genes, the SCE object needs to be
loaded:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"processed_data/sce.rds"</span><span class="op">)</span></span></code></pre></div>
<p>Next, edge directionality is calculated with the function
<tt>addDirectionality</tt>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">consensus_network</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addDirectionality.html">addDirectionality</a></span><span class="op">(</span><span class="va">consensus_network</span>, <span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<table class="table table" style="font-size: 13px; margin-left: auto; margin-right: auto;">
<caption style="font-size: initial !important;">
<span style="font-size:13px">A few example edges from the consensus
network after adding the direction of correlation</span>
</caption>
<thead><tr>
<th style="text-align:center;">
from
</th>
<th style="text-align:center;">
to
</th>
<th style="text-align:center;">
weight
</th>
<th style="text-align:center;">
n_supporting_replicates
</th>
<th style="text-align:center;">
supporting_replicates
</th>
<th style="text-align:center;">
rho
</th>
<th style="text-align:center;">
p.adj
</th>
<th style="text-align:center;">
direction
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ACAT2
</td>
<td style="text-align:center;">
0.0000426
</td>
<td style="text-align:center;">
3
</td>
<td style="text-align:center;">
H3c1,C2c1,C2c2
</td>
<td style="text-align:center;">
0.0293
</td>
<td style="text-align:center;">
0.20
</td>
<td style="text-align:center;">
+
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ACBD4
</td>
<td style="text-align:center;">
0.0000145
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
C2c2
</td>
<td style="text-align:center;">
-0.0202
</td>
<td style="text-align:center;">
0.46
</td>
<td style="text-align:center;">
-
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ACBD6
</td>
<td style="text-align:center;">
0.0011186
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
C1c1
</td>
<td style="text-align:center;">
0.0063
</td>
<td style="text-align:center;">
0.83
</td>
<td style="text-align:center;">
+
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ACBD7
</td>
<td style="text-align:center;">
0.0000081
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
C1c1
</td>
<td style="text-align:center;">
-0.0109
</td>
<td style="text-align:center;">
0.75
</td>
<td style="text-align:center;">
-
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ACER3
</td>
<td style="text-align:center;">
0.0000665
</td>
<td style="text-align:center;">
2
</td>
<td style="text-align:center;">
H1c1,H3c1
</td>
<td style="text-align:center;">
0.0170
</td>
<td style="text-align:center;">
0.54
</td>
<td style="text-align:center;">
+
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ACIN1
</td>
<td style="text-align:center;">
0.0001691
</td>
<td style="text-align:center;">
2
</td>
<td style="text-align:center;">
G1c2,C1c1
</td>
<td style="text-align:center;">
0.0234
</td>
<td style="text-align:center;">
0.35
</td>
<td style="text-align:center;">
+
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ACLY
</td>
<td style="text-align:center;">
0.0000402
</td>
<td style="text-align:center;">
2
</td>
<td style="text-align:center;">
H3c1,C2c1
</td>
<td style="text-align:center;">
0.0442
</td>
<td style="text-align:center;">
0.04
</td>
<td style="text-align:center;">
+
</td>
</tr>
<tr>
<td style="text-align:center;">
A1BG
</td>
<td style="text-align:center;">
ACO1
</td>
<td style="text-align:center;">
0.0000244
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
C2c1
</td>
<td style="text-align:center;">
0.0221
</td>
<td style="text-align:center;">
0.43
</td>
<td style="text-align:center;">
+
</td>
</tr>
</tbody>
</table>
<p><br><br></p>
</div>
</div>
<div class="section level2">
<h2 id="module-assignment">Module assignment<a class="anchor" aria-label="anchor" href="#module-assignment"></a>
</h2>
<p>Once the consensus network is calculated, it can be used to assign
co-expression modules jointly for all species. The center of each module
is a transcriptional regulator and the modules are assigned in 2 main
steps: 1) large initial modules are created by selecting a fixed number
of target genes per regulator, and 2) the initial modules are pruned
using a dynamic filtering approach to keep only the best targets of each
regulator.</p>
<div class="section level3">
<h3 id="selecting-key-transcriptional-regulators">Selecting key transcriptional regulators<a class="anchor" aria-label="anchor" href="#selecting-key-transcriptional-regulators"></a>
</h3>
<p>First, the transcriptional regulators that form the cores of the
co-expression modules have to be selected. To identify regulators that
are relevant for the biological process under study, it is a useful
approach to combine prior biological knowledge with information derived
from the dataset. This step is implemented by the function
<tt>getRegulators</tt>.</p>
<p>As for the prior biological knowledge, we recommend to select only
transcriptional regulators with known binding motifs based on motif
databases. To get this information, the user can choose the following
databases: JASPAR 2024 vertebrate core <span class="citation">[14]</span>, JASPAR 2024 unvalidated <span class="citation">[14]</span>, the IMAGE database <span class="citation">[15]</span>, or any combination of the above.
Alternatively, the user can provide a custom list of regulators selected
by their preferred method.</p>
<p>As for the information from the dataset, we recommend to select only
transcriptional regulators that are among the highly variable genes in
the data. For defining highly variable genes, <tt>getRegulators</tt>
relies on <tt>scran::getTopHVGs</tt> <span class="citation">[3]</span>.
Briefly, for each species a trend is fitted between the variance and
mean of the log-expression values across all genes, and the fitted value
of a gene is regarded as the technical component of variation, while the
residual from the trend is regarded as the biological component of
variation. All genes with a positive biological component are selected
as highly variable genes in each species (alternatively, a more
stringent variance cutoff, an adjusted p-value cutoff, a fixed number of
genes, or a fixed percentage of genes can also be defined). As the final
set of highly variable genes, the union of highly variable genes is
taken across species.</p>
<p>In case of the example dataset, we used transcriptional regulators
that 1) had an annotated motif in any of the 3 motif databases (JASPAR
2024 vertebrate core, JASPAR 2024 unvalidated and IMAGE) and 2) had a
positive biological component of variance in any of the species:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">regulators</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getRegulators.html">getRegulators</a></span><span class="op">(</span><span class="va">sce</span>, </span>
<span>                            source <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"jaspar_core"</span>, <span class="st">"jaspar_unvalidated"</span>, <span class="st">"image"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>This gave us 836 transcriptional regulators that included well-known
pluripotency factors and early neural regulators:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">regulators</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 836</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NANOG"</span>, <span class="st">"POU5F1"</span>, <span class="st">"PAX6"</span>, <span class="st">"NEUROD4"</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">regulators</span></span>
<span><span class="co">#&gt; [1] TRUE TRUE TRUE TRUE</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="assigning-initial-modules">Assigning initial modules<a class="anchor" aria-label="anchor" href="#assigning-initial-modules"></a>
</h3>
<p>Once the regulators that provide the starting point of the module
assignment are selected, an initial module can be defined around each of
them using the function <tt>assignInitialModules</tt>. Each of these
modules contain the regulator and its N best target genes (N is
recommended to be between 1000 and 5000). When choosing the best
targets, the genes are ranked based on how strongly they are connected
to the regulator (regulator-target adjacency).</p>
<p>In case of the example dataset, we defined 4000 as the number of
genes in the initial modules:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">initial_modules</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/assignInitialModules.html">assignInitialModules</a></span><span class="op">(</span><span class="va">consensus_network</span>, </span>
<span>                                        regulators <span class="op">=</span> <span class="va">regulators</span>, </span>
<span>                                        N <span class="op">=</span> <span class="fl">4000</span><span class="op">)</span></span></code></pre></div>
<table class="table table" style="font-size: 13px; margin-left: auto; margin-right: auto;">
<caption style="font-size: initial !important;">
<span style="font-size:13px">Part of the POU5F1 initial module</span>
</caption>
<thead><tr>
<th style="text-align:center;">
regulator
</th>
<th style="text-align:center;">
target
</th>
<th style="text-align:center;">
weight
</th>
<th style="text-align:center;">
n_supporting_replicates
</th>
<th style="text-align:center;">
supporting_replicates
</th>
<th style="text-align:center;">
rho
</th>
<th style="text-align:center;">
p.adj
</th>
<th style="text-align:center;">
direction
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
ACAT2
</td>
<td style="text-align:center;">
0.00080
</td>
<td style="text-align:center;">
4
</td>
<td style="text-align:center;width: 3cm; ">
H1c1,G1c1,C2c1,C2c2
</td>
<td style="text-align:center;">
0.125
</td>
<td style="text-align:center;">
1e-12
</td>
<td style="text-align:center;">
+
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
ACCS
</td>
<td style="text-align:center;">
0.00016
</td>
<td style="text-align:center;">
4
</td>
<td style="text-align:center;width: 3cm; ">
G1c1,G1c2,C1c2,C2c1
</td>
<td style="text-align:center;">
0.092
</td>
<td style="text-align:center;">
5e-06
</td>
<td style="text-align:center;">
+
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
ACER3
</td>
<td style="text-align:center;">
0.00053
</td>
<td style="text-align:center;">
3
</td>
<td style="text-align:center;width: 3cm; ">
H1c1,G1c1,C1c2
</td>
<td style="text-align:center;">
0.090
</td>
<td style="text-align:center;">
1e-06
</td>
<td style="text-align:center;">
+
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
ACLY
</td>
<td style="text-align:center;">
0.00081
</td>
<td style="text-align:center;">
4
</td>
<td style="text-align:center;width: 3cm; ">
H1c1,G1c1,C2c1,C2c2
</td>
<td style="text-align:center;">
0.189
</td>
<td style="text-align:center;">
2e-27
</td>
<td style="text-align:center;">
+
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
ACOT13
</td>
<td style="text-align:center;">
0.00300
</td>
<td style="text-align:center;">
6
</td>
<td style="text-align:center;width: 3cm; ">
H1c1,H3c1,G1c1,G1c2, C1c2,C2c2
</td>
<td style="text-align:center;">
0.086
</td>
<td style="text-align:center;">
7e-08
</td>
<td style="text-align:center;">
+
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
ACOT7
</td>
<td style="text-align:center;">
0.00475
</td>
<td style="text-align:center;">
8
</td>
<td style="text-align:center;width: 3cm; ">
H1c1,H2c1,G1c1,G1c2, C1c1,C1c2,C2c1,C2c2
</td>
<td style="text-align:center;">
0.337
</td>
<td style="text-align:center;">
7e-96
</td>
<td style="text-align:center;">
+
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
ACOT8
</td>
<td style="text-align:center;">
0.00083
</td>
<td style="text-align:center;">
3
</td>
<td style="text-align:center;width: 3cm; ">
G1c1,C1c1,C2c2
</td>
<td style="text-align:center;">
0.212
</td>
<td style="text-align:center;">
3e-35
</td>
<td style="text-align:center;">
+
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
ACP1
</td>
<td style="text-align:center;">
0.00053
</td>
<td style="text-align:center;">
3
</td>
<td style="text-align:center;width: 3cm; ">
G1c2,C1c2,C2c1
</td>
<td style="text-align:center;">
0.113
</td>
<td style="text-align:center;">
9e-11
</td>
<td style="text-align:center;">
+
</td>
</tr>
</tbody>
</table>
<p>The resulting modules are obviously too large at this point to be
biologically meaningful. The purpose of this step is rather to select a
uniformly sized pool of genes for all modules to perform the pruning
step on.</p>
</div>
<div class="section level3">
<h3 id="module-pruning">Module pruning<a class="anchor" aria-label="anchor" href="#module-pruning"></a>
</h3>
<p>The goal of pruning is to keep only the best targets of each
transcriptional regulator and thereby create biologically meaningful
modules that consist of co-regulated genes with shared cellular
functions.</p>
<p>Which targets are the best can be defined based on 2 metrics:</p>
<div class="figure" style="text-align: center">
<img src="adj_regulator_kIM.png" alt="Network concepts used for module pruning: regulator-target adjacency and intramodular connectivity." width="60%"><p class="caption">
Network concepts used for module pruning: regulator-target adjacency and
intramodular connectivity.
</p>
</div>
<ul>
<li>
<p><strong>Regulator-target adjacency</strong>
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>a</mi><mrow><mi mathvariant="normal">r</mi><mi mathvariant="normal">e</mi><mi mathvariant="normal">g</mi><mi mathvariant="normal">u</mi><mi mathvariant="normal">l</mi><mi mathvariant="normal">a</mi><mi mathvariant="normal">t</mi><mi mathvariant="normal">o</mi><mi mathvariant="normal">r</mi></mrow></msub><annotation encoding="application/x-tex">a_{\mathrm{regulator}}</annotation></semantics></math>):
The adjacency between a transcriptional regulator and a member gene in
its module. It ranges between 0 and 1 with a higher value representing a
stronger (activating or repressing) regulatory relationship.</p>
<ul>
<li><p><em>Pro</em>: Pruning based on the regulator-target adjacencies
selects the genes whose expression profiles co-vary the most with the
regulator’s expression profile.</p></li>
<li><p><em>Con</em>: Modules created based on the regulator-target
adjacencies alone might not fulfill the expectation of being densely
connected blocks of genes, apart from the link to the regulator, the
target genes often only have sparse/weak connections within the
module.</p></li>
</ul>
</li>
<li>
<p><strong>Intramodular connectivity</strong> (<em>kIM</em>): The
sum of adjacencies between a specific gene and all other members of the
module. If the module size is <em>n</em>, it ranges between 0 and
<em>n</em> - 1 with a higher value representing a gene that is more
strongly connected to the rest of the module.
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>k</mi><mi>I</mi><msubsup><mi>M</mi><mi>i</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>q</mi><mo stretchy="true" form="postfix">)</mo></mrow></msubsup><mo>=</mo><munder><mo>∑</mo><mtable><mtr><mtd columnalign="center" style="text-align: center"><mrow><mi>j</mi><mo>∈</mo><mi>q</mi></mrow></mtd></mtr><mtr><mtd columnalign="center" style="text-align: center"><mrow><mi>j</mi><mo>≠</mo><mi>i</mi></mrow></mtd></mtr></mtable></munder><msub><mi>a</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></mrow><annotation encoding="application/x-tex">kIM_i^{(q)} = \sum_{\substack{j \in q\\j \neq i}} a_{ij}</annotation></semantics></math></p>
<ul>
<li><p><em>Pro</em>: Pruning based on intramodular connectivities in
addition to the regulator-target adjacencies ensures that the chosen
targets co-vary not just with the main regulator but also with the rest
of the module. These intramodular connections between targets can carry
important information that are not captured by the links between
regulator and targets alone (e.g. combinatorial regulation).</p></li>
<li><p><em>Con</em>: Genes that have a generally high connectivity but
do not have a particularly strong regulatory relationship with the
transcriptional regulator might end up in the final module.</p></li>
</ul>
</li>
</ul>
<p>The initial targets can be filtered based one or both of these
metrics using 2 approaches:</p>
<ul>
<li>
<p><strong>topN</strong>: Keeps a fixed number of targets (default:
50) per regulator.</p>
<ul>
<li><p><em>Pro</em>: Quick, simple and widespread.</p></li>
<li><p><em>Con</em>: The approach assigns the same number of target
genes to each regulator even though biologically speaking different
regulators can have an effect on different numbers of genes. The number
of targets is in most cases arbitrary and does not take into account the
structure of the data at hand.</p></li>
</ul>
</li>
<li>
<p><strong>UIK</strong>: Applies dynamic stepwise pruning using
knee-point detection. In each step, the cumulative sum curve based on
the metric of choice is calculated per module, the knee point of the
curve is identified using the Unit Invariant Knee (UIK) method, then
only the targets that rank higher than the knee point are kept. The
modules containing less target genes than a minimum module size
(default: 20) are removed after each pruning step. The steps continue
until the percentage of removed modules becomes too high (default:
&gt;2%).</p>
<ul>
<li><p><em>Pro</em>: While setting a minimum module size prevents the
modules from becoming too small, the exact number of target genes per
regulator does not have to be pre-defined, in line with the notion that
different regulators can have an effect on different numbers of genes.
There are also no hard cutoffs applied to the regulator-target
adjacencies or intramodular connectivities, but by using knee point
detection the target genes are filtered in a data-driven way.</p></li>
<li><p><em>Con</em>: The approach still requires a predefined minimum
module size.</p></li>
</ul>
</li>
</ul>
<p>Compared to clustering-based module assignment approaches, both
pruning methods have the advantage that the modules are allowed to
overlap, and in addition to having its own module, a regulator can be
assigned to another regulator’s module as well. This is in line with the
notion that genes can fulfill several different functions in the cell
and gene regulation can be combinatorial.</p>
<p>In case of both pruning methods, when setting the fixed module size
or the threshold for the minimum module size not just biological but
also technical aspects have to be taken into consideration: for modules
smaller than ~20 genes, the correlation-based preservation statistics in
the next steps might be coupled with high uncertainty, therefore they
should be avoided.</p>
<p>Using the metrics and approaches described above, we implemented 3
methods for module pruning:</p>
<ul>
<li><p><strong>topN</strong>: Takes a fixed number of targets per
regulator with the highest regulator-target adjacencies.</p></li>
<li><p><strong>UIK_adj</strong>: Applies a dynamic stepwise pruning
based on the regulator-target adjacencies.</p></li>
<li><p><strong>UIK_adj_kIM</strong>: Applies a dynamic stepwise pruning
based on the regulator-target adjacencies and intramodular
connectivities alternately.</p></li>
</ul>
<p>In case of the example dataset, we used the method “UIK_adj_kIM”:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pruned_modules</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pruneModules.html">pruneModules</a></span><span class="op">(</span><span class="va">initial_modules</span>, </span>
<span>                               method <span class="op">=</span> <span class="st">"UIK_adj_kIM"</span>, </span>
<span>                               consensus_network <span class="op">=</span> <span class="va">consensus_network</span><span class="op">)</span></span></code></pre></div>
<p>The following figure shows the steps of the pruning process for the
POU5F1 module:</p>
<div class="figure" style="text-align: center">
<img src="POU5F1_detailed_UIK_adj_kIM.png" alt="Stepwise dynamic pruning approach exemplified by the POU5F1 module. In each step, we calculated the cumulative sum curve of the regulator-target adjacency or intramodular connectivitiy of the target genes alternately, then kept the targets that fell below the knee point of the curve. We continued this process for 3 iterations - until the module sizes became as small as possible without falling below 20 genes." width="55%"><p class="caption">
Stepwise dynamic pruning approach exemplified by the POU5F1 module. In
each step, we calculated the cumulative sum curve of the
regulator-target adjacency or intramodular connectivitiy of the target
genes alternately, then kept the targets that fell below the knee point
of the curve. We continued this process for 3 iterations - until the
module sizes became as small as possible without falling below 20 genes.
</p>
</div>
<table class="table table" style="font-size: 13px; margin-left: auto; margin-right: auto;">
<caption style="font-size: initial !important;">
<span style="font-size:13px">Part of the POU5F1 pruned module</span>
</caption>
<thead><tr>
<th style="text-align:center;">
regulator
</th>
<th style="text-align:center;">
target
</th>
<th style="text-align:center;">
weight
</th>
<th style="text-align:center;">
kIM
</th>
<th style="text-align:center;">
n_supporting_ replicates
</th>
<th style="text-align:center;">
supporting_ replicates
</th>
<th style="text-align:center;">
rho
</th>
<th style="text-align:center;">
p.adj
</th>
<th style="text-align:center;">
direction
</th>
<th style="text-align:center;">
module_size
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
AP1S2
</td>
<td style="text-align:center;">
0.026
</td>
<td style="text-align:center;">
0.4140710
</td>
<td style="text-align:center;width: 2.5cm; ">
8
</td>
<td style="text-align:center;">
H1c1,H2c1,H3c1, G1c2,C1c1,C1c2, C2c1,C2c2
</td>
<td style="text-align:center;">
0.38
</td>
<td style="text-align:center;">
3e-127
</td>
<td style="text-align:center;">
+
</td>
<td style="text-align:center;">
46
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
APOE
</td>
<td style="text-align:center;">
0.075
</td>
<td style="text-align:center;">
1.1619107
</td>
<td style="text-align:center;width: 2.5cm; ">
9
</td>
<td style="text-align:center;">
H1c1,H2c1,H3c1, G1c1,G1c2,C1c1, C1c2,C2c1,C2c2
</td>
<td style="text-align:center;">
0.58
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
+
</td>
<td style="text-align:center;">
46
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
CD24
</td>
<td style="text-align:center;">
0.143
</td>
<td style="text-align:center;">
1.9220911
</td>
<td style="text-align:center;width: 2.5cm; ">
9
</td>
<td style="text-align:center;">
H1c1,H2c1,H3c1, G1c1,G1c2,C1c1, C1c2,C2c1,C2c2
</td>
<td style="text-align:center;">
0.63
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
+
</td>
<td style="text-align:center;">
46
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
CDH2
</td>
<td style="text-align:center;">
0.036
</td>
<td style="text-align:center;">
0.6176581
</td>
<td style="text-align:center;width: 2.5cm; ">
8
</td>
<td style="text-align:center;">
H1c1,H2c1,H3c1, G1c1,G1c2,C1c1, C1c2,C2c2
</td>
<td style="text-align:center;">
-0.46
</td>
<td style="text-align:center;">
1e-194
</td>
<td style="text-align:center;">
-
</td>
<td style="text-align:center;">
46
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
CENPV
</td>
<td style="text-align:center;">
0.025
</td>
<td style="text-align:center;">
0.4156734
</td>
<td style="text-align:center;width: 2.5cm; ">
8
</td>
<td style="text-align:center;">
H1c1,H3c1,G1c1, G1c2,C1c1,C1c2, C2c1,C2c2
</td>
<td style="text-align:center;">
-0.33
</td>
<td style="text-align:center;">
6e-93
</td>
<td style="text-align:center;">
-
</td>
<td style="text-align:center;">
46
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
CHCHD2
</td>
<td style="text-align:center;">
0.031
</td>
<td style="text-align:center;">
0.5833867
</td>
<td style="text-align:center;width: 2.5cm; ">
7
</td>
<td style="text-align:center;">
H2c1,H3c1,G1c2, C1c1,C1c2,C2c1, C2c2
</td>
<td style="text-align:center;">
0.39
</td>
<td style="text-align:center;">
8e-135
</td>
<td style="text-align:center;">
+
</td>
<td style="text-align:center;">
46
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
CLDN6
</td>
<td style="text-align:center;">
0.033
</td>
<td style="text-align:center;">
0.5877413
</td>
<td style="text-align:center;width: 2.5cm; ">
9
</td>
<td style="text-align:center;">
H1c1,H2c1,H3c1, G1c1,G1c2,C1c1, C1c2,C2c1,C2c2
</td>
<td style="text-align:center;">
0.57
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
+
</td>
<td style="text-align:center;">
46
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
COX7C
</td>
<td style="text-align:center;">
0.097
</td>
<td style="text-align:center;">
1.9427046
</td>
<td style="text-align:center;width: 2.5cm; ">
9
</td>
<td style="text-align:center;">
H1c1,H2c1,H3c1, G1c1,G1c2,C1c1, C1c2,C2c1,C2c2
</td>
<td style="text-align:center;">
0.58
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
+
</td>
<td style="text-align:center;">
46
</td>
</tr>
</tbody>
</table>
<p>Since we use the “UIK_adj_kIM” approach, the final module sizes are
not predefined but shaped by the data. For this particular dataset, the
number of genes in the pruned modules range from 25 to 63 with a median
of 45:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotModuleSizeDistribution.html">plotModuleSizeDistribution</a></span><span class="op">(</span><span class="va">pruned_modules</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="module_size_distribution.png" alt="Distribution of module sizes after pruning. Median size: 45." width="48%"><p class="caption">
Distribution of module sizes after pruning. Median size: 45.
</p>
</div>
</div>
<div class="section level3">
<h3 id="visualizing-modules">Visualizing modules<a class="anchor" aria-label="anchor" href="#visualizing-modules"></a>
</h3>
<p>The most important connections of the modules can be plotted using
the function <tt>plotNetworks</tt>. For each module, the resulting plots
show the strongest intramodular connections based on the consensus
network and the genes involved in these connections. The following plot
depicts the top 300 connections of 6 example modules, colored by the
direction of correlation:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_modules</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NANOG"</span>, <span class="st">"POU5F1"</span>, <span class="st">"SALL4"</span>,  <span class="st">"NEUROD4"</span>, <span class="st">"PAX6"</span>, <span class="st">"FEZF2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotNetworks.html">plotNetworks</a></span><span class="op">(</span><span class="va">example_modules</span>, </span>
<span>             <span class="va">pruned_modules</span>, </span>
<span>             <span class="va">consensus_network</span>, </span>
<span>             color_by <span class="op">=</span> <span class="st">"direction"</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="network_examples.png" alt="The 300 strongest connections of 6 example modules. The thickness of the edges represents the consensus edge weights and the color of the edges represents the direction of interaction (positively/negatively correlated)." width="100%"><p class="caption">
The 300 strongest connections of 6 example modules. The thickness of the
edges represents the consensus edge weights and the color of the edges
represents the direction of interaction (positively/negatively
correlated).
</p>
</div>
</div>
<div class="section level3">
<h3 id="creating-random-modules">Creating random modules<a class="anchor" aria-label="anchor" href="#creating-random-modules"></a>
</h3>
<p>Along with the pruned modules, there is an option to create random
modules as well using the function <tt>createRandomModules</tt>. The
random modules have the same regulators and contain the same number of
target genes as the actual pruned modules, but these target genes are
randomly drawn from all network genes.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">random_modules</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createRandomModules.html">createRandomModules</a></span><span class="op">(</span><span class="va">pruned_modules</span>, </span>
<span>                                      network_genes <span class="op">=</span> <span class="va">genes</span><span class="op">)</span></span></code></pre></div>
<p>In the next steps of the pipeline, the actual modules are compared to
these random modules in terms of various statistics to check whether the
two groups of modules behave in general differently and to remove those
individual actual modules that show too similar characteristics to the
random modules.</p>
<p><br><br></p>
</div>
</div>
<div class="section level2">
<h2 id="characterizing-modules">Characterizing modules<a class="anchor" aria-label="anchor" href="#characterizing-modules"></a>
</h2>
<p>Since co-expression modules are often more closely associated with a
cellular function than any particular gene alone, it is informative to
investigate which cell types the different modules are active in and
which pathways they are associated to. This can be achieved for example
by calculating module eigengenes and peforming gene set enrichment
analysis.</p>
<div class="section level3">
<h3 id="calculating-module-eigengenes">Calculating module eigengenes<a class="anchor" aria-label="anchor" href="#calculating-module-eigengenes"></a>
</h3>
<p>A concept adapted from WGCNA <span class="citation">[16]</span>, a
module eigengene summarizes the expression profile of an entire module,
and it is calculated as the first principal component of the module
expression data. Effectively, it is a weighted mean of the individual
genes’ expression profiles. In <tt>CroConet</tt>, the calculation of
eigengene profiles is implemented by <tt>calculateEigengenes</tt>. This
function outputs other metrics as well for representing the expression
patterns of a module, namely the mean expression across all module
members and the expression of the regulator.</p>
<p>If a module contains both activated and repressed targets of the
transcriptional regulator, calculating the eigengene (or any other
summary expression profiles) across both directions of regulation does
not make biological sense and leads to the dilution of signal. It makes
more sense to calculate the eigengene either for the activated targets
only (<tt>direction_of_regulation</tt> = “+_only”, the default) or for
the activated and repressed targets separately
(<tt>direction_of_regulation</tt> = “+-_separately”). If an eigengene
across all targets is desired irrespective of the direction of
regulation, <tt>direction_of_regulation</tt> should be set to
“all_together”.</p>
<p>In case of the example dataset, we calculated module eigengenes for
the activated targets of each module:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">eigengenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateEigengenes.html">calculateEigengenes</a></span><span class="op">(</span><span class="va">regulators</span>,</span>
<span>                                  <span class="va">pruned_modules</span>, </span>
<span>                                  <span class="va">sce</span>,</span>
<span>                                  direction_of_regulation <span class="op">=</span> <span class="st">"+_only"</span><span class="op">)</span></span></code></pre></div>
<table class="table table" style="font-size: 12.9px; margin-left: auto; margin-right: auto;">
<caption style="font-size: initial !important;">
<span style="font-size:13px">Eigengene profile of the activated POU5F1
module across a couple of example cells</span>
</caption>
<thead><tr>
<th style="text-align:left;">
cell
</th>
<th style="text-align:left;">
species
</th>
<th style="text-align:right;">
pseudotime
</th>
<th style="text-align:left;">
cell_type
</th>
<th style="text-align:left;">
module
</th>
<th style="text-align:right;">
eigengene
</th>
<th style="text-align:right;">
mean_expr
</th>
<th style="text-align:right;">
regulator_expr
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
ATCTCAGGAAAACT
</td>
<td style="text-align:left;">
human
</td>
<td style="text-align:right;">
0.504
</td>
<td style="text-align:left;">
Early_Ectoderm
</td>
<td style="text-align:left;">
POU5F1(+)
</td>
<td style="text-align:right;">
-0.008
</td>
<td style="text-align:right;">
-0.317
</td>
<td style="text-align:right;">
0.242
</td>
</tr>
<tr>
<td style="text-align:left;">
ATCTCAGGAAAGTT
</td>
<td style="text-align:left;">
human
</td>
<td style="text-align:right;">
0.604
</td>
<td style="text-align:left;">
Early_Ectoderm
</td>
<td style="text-align:left;">
POU5F1(+)
</td>
<td style="text-align:right;">
-0.013
</td>
<td style="text-align:right;">
-0.529
</td>
<td style="text-align:right;">
-0.760
</td>
</tr>
<tr>
<td style="text-align:left;">
ATCTCAGGAAATTG
</td>
<td style="text-align:left;">
human
</td>
<td style="text-align:right;">
0.570
</td>
<td style="text-align:left;">
Early_Ectoderm
</td>
<td style="text-align:left;">
POU5F1(+)
</td>
<td style="text-align:right;">
-0.011
</td>
<td style="text-align:right;">
-0.422
</td>
<td style="text-align:right;">
-0.015
</td>
</tr>
<tr>
<td style="text-align:left;">
ATCTCAGGAAGATT
</td>
<td style="text-align:left;">
human
</td>
<td style="text-align:right;">
0.568
</td>
<td style="text-align:left;">
Early_Ectoderm
</td>
<td style="text-align:left;">
POU5F1(+)
</td>
<td style="text-align:right;">
-0.009
</td>
<td style="text-align:right;">
-0.347
</td>
<td style="text-align:right;">
-0.760
</td>
</tr>
<tr>
<td style="text-align:left;">
ATCTCAGGAATACA
</td>
<td style="text-align:left;">
human
</td>
<td style="text-align:right;">
0.525
</td>
<td style="text-align:left;">
Early_Ectoderm
</td>
<td style="text-align:left;">
POU5F1(+)
</td>
<td style="text-align:right;">
-0.008
</td>
<td style="text-align:right;">
-0.275
</td>
<td style="text-align:right;">
-0.760
</td>
</tr>
<tr>
<td style="text-align:left;">
ATCTCAGGAATCTT
</td>
<td style="text-align:left;">
human
</td>
<td style="text-align:right;">
0.515
</td>
<td style="text-align:left;">
Early_Ectoderm
</td>
<td style="text-align:left;">
POU5F1(+)
</td>
<td style="text-align:right;">
-0.008
</td>
<td style="text-align:right;">
-0.286
</td>
<td style="text-align:right;">
-0.760
</td>
</tr>
<tr>
<td style="text-align:left;">
ATCTCAGGACAATA
</td>
<td style="text-align:left;">
human
</td>
<td style="text-align:right;">
0.576
</td>
<td style="text-align:left;">
Early_Ectoderm
</td>
<td style="text-align:left;">
POU5F1(+)
</td>
<td style="text-align:right;">
-0.010
</td>
<td style="text-align:right;">
-0.405
</td>
<td style="text-align:right;">
-0.760
</td>
</tr>
<tr>
<td style="text-align:left;">
ATCTCAGGACCGGC
</td>
<td style="text-align:left;">
human
</td>
<td style="text-align:right;">
0.525
</td>
<td style="text-align:left;">
Early_Ectoderm
</td>
<td style="text-align:left;">
POU5F1(+)
</td>
<td style="text-align:right;">
-0.005
</td>
<td style="text-align:right;">
-0.190
</td>
<td style="text-align:right;">
-0.760
</td>
</tr>
</tbody>
</table>
<p>The eigengene profiles can then be plotted as a heatmap using
<tt>plotEigengeneHeatmap</tt>. Here we focus on the 6 example modules
again, that contain 3 pluripotency regulators (NANOG, POU5F1 and CREBL3)
and 3 neural regulators (PAX6, NEUROD4 and FEZF2):</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">eigengenes_examples</span> <span class="op">&lt;-</span> <span class="va">eigengenes</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">module</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">example_modules</span>, <span class="st">"(+)"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>module <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">module</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">example_modules</span>, <span class="st">"(+)"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotEigengeneHeatmap.html">plotEigengeneHeatmap</a></span><span class="op">(</span><span class="va">eigengenes_examples</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="eigengenes_examples.png" alt="The eigengenes of the example modules." width="70%"><p class="caption">
The eigengenes of the example modules.
</p>
</div>
<p>As expected, the activated targets of the pluripotency regulators
show the highest expression levels at early pseudotime stages and become
downregulated later on, while the activated targets of the neural
regulators have low expression levels at first, and become upregulated
towards later stages of the differentiation process.</p>
</div>
<div class="section level3">
<h3 id="gene-set-enrichment-analysis">Gene set enrichment analysis<a class="anchor" aria-label="anchor" href="#gene-set-enrichment-analysis"></a>
</h3>
<p>Gene set enrichment analysis offers useful information for the
functional annotation of the modules. Various databases and tools can be
used for such an analysis, including Gene Ontology (GO) <span class="citation">[17, 18]</span> and the related R package
<tt>topGO</tt> <span class="citation">[19]</span>, the Reactome Pathway
Database <span class="citation">[20]</span> and the R package
<tt>ReactomePA</tt> <span class="citation">[21]</span>, or STRING <span class="citation">[22]</span> and the R package <tt>STRINGdb</tt> <span class="citation">[22]</span>. It is also worth mentioning
<tt>clusterProfiler</tt> <span class="citation">[23]</span> which
provides a universal interface for enrichment analysis based on various
sources and makes it easy to analyze and compare several different gene
sets in a single run.</p>
<p>In the next steps, we perform pathway enrichment analysis on the 6
example modules using the Reactome Pathway Database. First, a data frame
containing the activated targets of the example modules is prepared and
the gene names are converted to ENTREZ IDs using the conversion table
stored in “sym2enrtez.rds” (this was created based on the <a href="https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/gencode.v32.metadata.EntrezGene.gz" class="external-link">GENCODE
“Entrez gene ids” metadata file</a> corresponding to the human genome
version used for mapping).</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sym2entrez</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"RDS/sym2entrez.rds"</span><span class="op">)</span></span>
<span><span class="va">example_modules_4_reactome</span> <span class="op">&lt;-</span> <span class="va">pruned_modules</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">regulator</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">example_modules</span> <span class="op">&amp;</span> <span class="va">direction</span> <span class="op">==</span> <span class="st">"+"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>regulator <span class="op">=</span> <span class="va">example_modules</span>,</span>
<span>                              target <span class="op">=</span> <span class="va">example_modules</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">inner_join</a></span><span class="op">(</span><span class="va">sym2entrez</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"target"</span> <span class="op">=</span> <span class="st">"gene_name"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>regulator <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">regulator</span>, <span class="va">example_modules</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Next, the enrichment analysis is done using the wrapper function
<tt>clusterProfiler::compareCluster</tt> and the testing implemented by
<tt>ReactomePA::enrichPathway</tt>. The universe is defined as all
network genes.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">enrich_res</span> <span class="op">&lt;-</span> <span class="fu">clusterProfiler</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/compareCluster.html" class="external-link">compareCluster</a></span><span class="op">(</span><span class="va">entrez_id</span> <span class="op">~</span> <span class="va">regulator</span>,</span>
<span>                                              <span class="st">"enrichPathway"</span>,</span>
<span>                                              <span class="va">example_modules_4_reactome</span>,</span>
<span>                                              universe <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">sym2entrez</span><span class="op">$</span><span class="va">entrez_id</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Finally, the results are plotted for all modules together:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">clusterProfiler</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/dotplot.html" class="external-link">dotplot</a></span><span class="op">(</span><span class="va">enrich_res</span>, showCategory <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="Reactome_dotplot.png" alt="Reactome enrichment analysis on 6 example modules." width="95%"><p class="caption">
Reactome enrichment analysis on 6 example modules.
</p>
</div>
<p>Encouragingly, the modules of the pluripotency regulators show
enrichment for several pathways related to pluripotency (e.g. <em>POU5F1
(OCT4), SOX2, NANOG activate genes related to proliferation</em> and
<em>Transcriptional regulation of pluripotent stem cells</em>), while
the modules of the neural regulators show enrichment for several
pathways related to the nerual lineage (e.g. <em>Axon guidance</em> and
<em>Nervous system development</em>).</p>
<p><br><br></p>
</div>
</div>
<div class="section level2">
<h2 id="module-preservation-within-and-across-species">Module preservation within and across species<a class="anchor" aria-label="anchor" href="#module-preservation-within-and-across-species"></a>
</h2>
<p>In the next part of the CroCoNet workflow, we quantify how
well-preseved the topologies of the modules are between different pairs
of replicates, both within and across species, and summarize these
preservation scores as neighbour-joining trees per module.</p>
<div class="section level3">
<h3 id="calculating-module-preservation-statistics">Calculating module preservation statistics<a class="anchor" aria-label="anchor" href="#calculating-module-preservation-statistics"></a>
</h3>
<p>For the calculation of preservation statistics, we use the joint
module assignment derived from the consensus network but compare
topological properties directly between the replicate-wise networks.
This way, a module is always defined as the same set of genes, but the
adjacencies and connectivities of these genes can differ from replicate
to replicate; poorly preserved modules are expected to have many, while
well-preserved modules are expected to have few such topological
differences.</p>
<p>Compared to the standard approach of cross-species module comparison
(defining modules for each species separately and then cross-tabulating
the module member genes across species), topology-based metrics can
detect way more subtle differences. In addition, the results of
cross-tabulation are strongly affected by statistical power: if we find
a low overlap across species, it is difficult to disentangle to what
extent this is due to real biological differences and to what extent
this is just due to having a low power for module detection in the first
place. In contrast, a comparison of network topology for a joint module
assignment is much less dependent on the power.</p>
<p>As part of the function <tt>calculatePresStats</tt>, two preservation
statistics (adapted from WGCNA <span class="citation">[24]</span>) are
implemented to quantify different aspects of this topological
comparison:</p>
<ul>
<li><p><strong>Correlation of adjacencies</strong> (cor_adj): the
correlation of all edge weights within the module in the network of
replicate1 VS in the network of replicate2
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>c</mi><mi>o</mi><mi>r</mi><mi>.</mi><mi>a</mi><mi>d</mi><msup><mi>j</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>q</mi><mo stretchy="true" form="postfix">)</mo></mrow></msup><mo>=</mo><mrow><mi mathvariant="normal">c</mi><mi mathvariant="normal">o</mi><mi mathvariant="normal">r</mi></mrow><mrow><mo stretchy="true" form="prefix">(</mo><msup><mi>a</mi><mrow><mrow><mo stretchy="true" form="prefix">[</mo><mi>n</mi><mi>e</mi><mi>t</mi><mn>1</mn><mo stretchy="true" form="postfix">]</mo></mrow><mrow><mo stretchy="true" form="prefix">(</mo><mi>q</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></msup><mo>,</mo><msup><mi>a</mi><mrow><mrow><mo stretchy="true" form="prefix">[</mo><mi>n</mi><mi>e</mi><mi>t</mi><mn>2</mn><mo stretchy="true" form="postfix">]</mo></mrow><mrow><mo stretchy="true" form="prefix">(</mo><mi>q</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></msup><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">cor.adj^{(q)} = \mathrm{cor}(a^{[net1](q)}, a^{[net2](q)})</annotation></semantics></math>
where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>a</mi><mrow><mrow><mo stretchy="true" form="prefix">[</mo><mi>n</mi><mi>e</mi><mi>t</mi><mo stretchy="true" form="postfix">]</mo></mrow><mrow><mo stretchy="true" form="prefix">(</mo><mi>q</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></msup><annotation encoding="application/x-tex">a^{[net](q)}</annotation></semantics></math>
is the vector of intramodular adjacencies in module
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>q</mi><annotation encoding="application/x-tex">q</annotation></semantics></math>
and network
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>n</mi><mi>e</mi><mi>t</mi></mrow><annotation encoding="application/x-tex">net</annotation></semantics></math>.</p></li>
<li><p><strong>Correlation of intramodular connectivities</strong>
(cor_kIM): the correlation of the intramodular connectivities per module
member gene in the network of replicate1 VS in the network of replicate2
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>c</mi><mi>o</mi><mi>r</mi><mi>.</mi><mi>k</mi><mi>I</mi><msup><mi>M</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>q</mi><mo stretchy="true" form="postfix">)</mo></mrow></msup><mo>=</mo><mrow><mi mathvariant="normal">c</mi><mi mathvariant="normal">o</mi><mi mathvariant="normal">r</mi></mrow><mrow><mo stretchy="true" form="prefix">(</mo><mi>k</mi><mi>I</mi><msup><mi>M</mi><mrow><mrow><mo stretchy="true" form="prefix">[</mo><mi>n</mi><mi>e</mi><mi>t</mi><mn>1</mn><mo stretchy="true" form="postfix">]</mo></mrow><mrow><mo stretchy="true" form="prefix">(</mo><mi>q</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></msup><mo>,</mo><mi>k</mi><mi>I</mi><msup><mi>M</mi><mrow><mrow><mo stretchy="true" form="prefix">[</mo><mi>n</mi><mi>e</mi><mi>t</mi><mn>2</mn><mo stretchy="true" form="postfix">]</mo></mrow><mrow><mo stretchy="true" form="prefix">(</mo><mi>q</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></msup><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">cor.kIM^{(q)} = \mathrm{cor}(kIM^{[net1](q)}, kIM^{[net2](q)})</annotation></semantics></math>
where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>k</mi><mi>I</mi><msup><mi>M</mi><mrow><mrow><mo stretchy="true" form="prefix">[</mo><mi>n</mi><mi>e</mi><mi>t</mi><mo stretchy="true" form="postfix">]</mo></mrow><mrow><mo stretchy="true" form="prefix">(</mo><mi>q</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></msup></mrow><annotation encoding="application/x-tex">kIM^{[net](q)}</annotation></semantics></math>
is the vector of intramodular connetivities per gene in module
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>q</mi><annotation encoding="application/x-tex">q</annotation></semantics></math>
and network
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>n</mi><mi>e</mi><mi>t</mi></mrow><annotation encoding="application/x-tex">net</annotation></semantics></math>.</p></li>
</ul>
<p>Both of these statistics can be calculated with module jackknifing
(by setting <tt>jackknife</tt> to TRUE). This means that the function
creates all possible jackknifed versions of each input module by
removing each target gene assigned to that module (the regulator itself
is never excluded), then it calculates the preservation statistics for
all of these jackknifed module versions in addition to the original
module. The advantage of using jackknfing is that a confidence interval
can be calculated for each module and statistic instead of just getting
a single value. Later on in the pipeline, jackknifing can also provide
information about which target genes within a conserved/diverged module
are particularly responsible for the conservation/divergence. If
jackknifing is not desired, the user can just set <tt>jackknife</tt> to
FALSE which will also substantially reduce running times.</p>
<p>In case of the example dataset, we calculated both preservation
statistics and opted for using jackknifing:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pres_stats_jk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculatePresStats.html">calculatePresStats</a></span><span class="op">(</span><span class="va">pruned_modules</span>, </span>
<span>                                    <span class="va">network_list</span>,</span>
<span>                                    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cor_adj"</span>, <span class="st">"cor_kIM"</span><span class="op">)</span>, </span>
<span>                                    <span class="va">replicate2species</span>,</span>
<span>                                    jackknifing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">random_pres_stats_jk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculatePresStats.html">calculatePresStats</a></span><span class="op">(</span><span class="va">random_modules</span>, </span>
<span>                                           <span class="va">network_list</span>,</span>
<span>                                           <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cor_adj"</span>, <span class="st">"cor_kIM"</span><span class="op">)</span>, </span>
<span>                                           <span class="va">replicate2species</span>,</span>
<span>                                           jackknifing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<table class="table table" style="font-size: 13px; margin-left: auto; margin-right: auto;">
<caption style="font-size: initial !important;">
<span style="font-size:13px">Preservation statistics of the POU5F1
module and one of its jackknifed module versions for a couple of example
replicate pairs</span>
</caption>
<thead><tr>
<th style="text-align:center;">
regulator
</th>
<th style="text-align:center;">
type
</th>
<th style="text-align:center;">
id
</th>
<th style="text-align:center;">
gene_ removed
</th>
<th style="text-align:center;">
replicate1
</th>
<th style="text-align:center;">
replicate2
</th>
<th style="text-align:center;">
species1
</th>
<th style="text-align:center;">
species2
</th>
<th style="text-align:center;">
cor_adj
</th>
<th style="text-align:center;">
cor_kIM
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
orig
</td>
<td style="text-align:center;">
POU5F1_orig
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
H1c1
</td>
<td style="text-align:center;">
H2c1
</td>
<td style="text-align:center;">
human
</td>
<td style="text-align:center;">
human
</td>
<td style="text-align:center;">
0.589
</td>
<td style="text-align:center;">
0.726
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
orig
</td>
<td style="text-align:center;">
POU5F1_orig
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
H1c1
</td>
<td style="text-align:center;">
H3c1
</td>
<td style="text-align:center;">
human
</td>
<td style="text-align:center;">
human
</td>
<td style="text-align:center;">
0.543
</td>
<td style="text-align:center;">
0.696
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
orig
</td>
<td style="text-align:center;">
POU5F1_orig
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
H1c1
</td>
<td style="text-align:center;">
G1c1
</td>
<td style="text-align:center;">
human
</td>
<td style="text-align:center;">
gorilla
</td>
<td style="text-align:center;">
0.474
</td>
<td style="text-align:center;">
0.644
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
orig
</td>
<td style="text-align:center;">
POU5F1_orig
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
H1c1
</td>
<td style="text-align:center;">
G1c2
</td>
<td style="text-align:center;">
human
</td>
<td style="text-align:center;">
gorilla
</td>
<td style="text-align:center;">
0.544
</td>
<td style="text-align:center;">
0.676
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
jk
</td>
<td style="text-align:center;">
POU5F1_jk_APOE
</td>
<td style="text-align:center;">
APOE
</td>
<td style="text-align:center;">
H1c1
</td>
<td style="text-align:center;">
H2c1
</td>
<td style="text-align:center;">
human
</td>
<td style="text-align:center;">
human
</td>
<td style="text-align:center;">
0.585
</td>
<td style="text-align:center;">
0.724
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
jk
</td>
<td style="text-align:center;">
POU5F1_jk_APOE
</td>
<td style="text-align:center;">
APOE
</td>
<td style="text-align:center;">
H1c1
</td>
<td style="text-align:center;">
H3c1
</td>
<td style="text-align:center;">
human
</td>
<td style="text-align:center;">
human
</td>
<td style="text-align:center;">
0.537
</td>
<td style="text-align:center;">
0.688
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
jk
</td>
<td style="text-align:center;">
POU5F1_jk_APOE
</td>
<td style="text-align:center;">
APOE
</td>
<td style="text-align:center;">
H1c1
</td>
<td style="text-align:center;">
G1c1
</td>
<td style="text-align:center;">
human
</td>
<td style="text-align:center;">
gorilla
</td>
<td style="text-align:center;">
0.489
</td>
<td style="text-align:center;">
0.666
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
jk
</td>
<td style="text-align:center;">
POU5F1_jk_APOE
</td>
<td style="text-align:center;">
APOE
</td>
<td style="text-align:center;">
H1c1
</td>
<td style="text-align:center;">
G1c2
</td>
<td style="text-align:center;">
human
</td>
<td style="text-align:center;">
gorilla
</td>
<td style="text-align:center;">
0.556
</td>
<td style="text-align:center;">
0.697
</td>
</tr>
</tbody>
</table>
<p>If the preservation statistics are calculated with jackknifing, a
statistic of interest can be estimated for a module as a whole by taking
its median across all jackknifed versions along with the confidence
interval if the median. This summarization is performed by the function
<tt>summarizeJackknifePresStats</tt>:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pres_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarizeJackknifeStats.html">summarizeJackknifeStats</a></span><span class="op">(</span><span class="va">pres_stats_jk</span><span class="op">)</span></span>
<span><span class="va">random_pres_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarizeJackknifeStats.html">summarizeJackknifeStats</a></span><span class="op">(</span><span class="va">random_pres_stats_jk</span><span class="op">)</span></span></code></pre></div>
<table class="table table" style="font-size: 13px; margin-left: auto; margin-right: auto;">
<caption style="font-size: initial !important;">
<span style="font-size:13px">Preservation statistics of the POU5F1
module for a couple of example replicate pairs, after summarizing the
values of all jackknifed module versions</span>
</caption>
<thead><tr>
<th style="text-align:center;">
regulator
</th>
<th style="text-align:center;">
replicate1
</th>
<th style="text-align:center;">
replicate2
</th>
<th style="text-align:center;">
species1
</th>
<th style="text-align:center;">
species2
</th>
<th style="text-align:center;">
cor_adj
</th>
<th style="text-align:center;">
var_cor_adj
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
H1c1
</td>
<td style="text-align:center;">
H2c1
</td>
<td style="text-align:center;">
human
</td>
<td style="text-align:center;">
human
</td>
<td style="text-align:center;">
0.590
</td>
<td style="text-align:center;">
0.000113
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
H1c1
</td>
<td style="text-align:center;">
H3c1
</td>
<td style="text-align:center;">
human
</td>
<td style="text-align:center;">
human
</td>
<td style="text-align:center;">
0.544
</td>
<td style="text-align:center;">
0.000154
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
H1c1
</td>
<td style="text-align:center;">
G1c1
</td>
<td style="text-align:center;">
human
</td>
<td style="text-align:center;">
gorilla
</td>
<td style="text-align:center;">
0.473
</td>
<td style="text-align:center;">
0.000251
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
H1c1
</td>
<td style="text-align:center;">
G1c2
</td>
<td style="text-align:center;">
human
</td>
<td style="text-align:center;">
gorilla
</td>
<td style="text-align:center;">
0.545
</td>
<td style="text-align:center;">
0.000236
</td>
</tr>
</tbody>
</table>
<table class="table table" style="font-size: 13px; margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:center;">
lwr_cor_adj
</th>
<th style="text-align:center;">
upr_cor_adj
</th>
<th style="text-align:center;">
cor_kIM
</th>
<th style="text-align:center;">
var_cor_kIM
</th>
<th style="text-align:center;">
lwr_cor_kIM
</th>
<th style="text-align:center;">
upr_cor_kIM
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:center;">
0.585
</td>
<td style="text-align:center;">
0.593
</td>
<td style="text-align:center;">
0.727
</td>
<td style="text-align:center;">
0.000224
</td>
<td style="text-align:center;">
0.724
</td>
<td style="text-align:center;">
0.731
</td>
</tr>
<tr>
<td style="text-align:center;">
0.539
</td>
<td style="text-align:center;">
0.546
</td>
<td style="text-align:center;">
0.695
</td>
<td style="text-align:center;">
0.000284
</td>
<td style="text-align:center;">
0.688
</td>
<td style="text-align:center;">
0.700
</td>
</tr>
<tr>
<td style="text-align:center;">
0.470
</td>
<td style="text-align:center;">
0.478
</td>
<td style="text-align:center;">
0.641
</td>
<td style="text-align:center;">
0.000327
</td>
<td style="text-align:center;">
0.638
</td>
<td style="text-align:center;">
0.646
</td>
</tr>
<tr>
<td style="text-align:center;">
0.540
</td>
<td style="text-align:center;">
0.547
</td>
<td style="text-align:center;">
0.673
</td>
<td style="text-align:center;">
0.000410
</td>
<td style="text-align:center;">
0.668
</td>
<td style="text-align:center;">
0.677
</td>
</tr>
</tbody>
</table>
<p>There are 3 easy-to-check expectations that a statistic has to
fulfill if it is a good measure of module preservation:</p>
<ul>
<li><p>The scores are expected to be higher for the actual modules than
for the random modules.</p></li>
<li><p>The scores are expected to be higher within species than across
species.</p></li>
<li><p>The difference between the within-species and cross-species
scores is expected to increase with increasing phylogenetic distance
between the two species compared.</p></li>
</ul>
<p>As a visual check, the function <tt>plotPresStatDistributions</tt>
plots the distributions of the preservation scores for the actual and
random modules and for within-species and cross-species replicate pairs,
whereas the function <tt>plotPresStats</tt> plots the cross-species
scores against the within-species scores per species pair:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotPresStatDistributions.html">plotPresStatDistributions</a></span><span class="op">(</span><span class="va">pres_stats</span>, </span>
<span>                          <span class="va">random_pres_stats</span>, </span>
<span>                          <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cor_adj"</span>, <span class="st">"cor_kIM"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="pres_stat_distributions.png" alt="Distribution of preservation statistics within and across species, for the actual and random modules." width="67%"><p class="caption">
Distribution of preservation statistics within and across species, for
the actual and random modules.
</p>
</div>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotPresStats.html">plotPresStats</a></span><span class="op">(</span><span class="va">pres_stats</span>, </span>
<span>              <span class="va">random_pres_stats</span>, </span>
<span>              <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cor_adj"</span>, <span class="st">"cor_kIM"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="pres_stat_across_vs_within.png" alt="Cross-species VS within-species preservation scores per species pair for the actual and random modules." width="100%"><p class="caption">
Cross-species VS within-species preservation scores per species pair for
the actual and random modules.
</p>
</div>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/comparePresStats.html">comparePresStats</a></span><span class="op">(</span><span class="va">pres_stats</span>, </span>
<span>                 <span class="va">random_pres_stats</span>, </span>
<span>                 <span class="va">tree</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="pres_stat_comparison.png" alt="Difference of the &lt;em&gt;cor.kIM&lt;/em&gt; / &lt;em&gt;cor.adj&lt;/em&gt; scores between each actual and the corresponding random module (left) and inverted Pearson's correlation (-&lt;em&gt;r&lt;/em&gt; ) between &lt;em&gt;cor.kIM&lt;/em&gt; / &lt;em&gt;cor.adj&lt;/em&gt; and phylogenetic distance (right)." width="75%"><p class="caption">
Difference of the <em>cor.kIM</em> / <em>cor.adj</em> scores between
each actual and the corresponding random module (left) and inverted
Pearson’s correlation (-<em>r</em> ) between <em>cor.kIM</em> /
<em>cor.adj</em> and phylogenetic distance (right).
</p>
</div>
<p>The plots show that when applied to the example dataset, both
<em>cor.kIM</em> and <em>cor.adj</em> fulfill all 3 expectations, but
since the differences between the actual and random modules as well as
between the within-species and cross-species comparisons tend to be
higher for <em>cor.kIM</em>, we use this statistic for the rest of the
analysis.</p>
</div>
<div class="section level3">
<h3 id="converting-preservation-scores-to-distance-measures">Converting preservation scores to distance measures<a class="anchor" aria-label="anchor" href="#converting-preservation-scores-to-distance-measures"></a>
</h3>
<p>As a preparation for tree reconstruction, the preservation statistics
need to be converted to distance measures. This is done by the function
<tt>convertPresToDist</tt> using the formula:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>d</mi><mo>=</mo><mfrac><mrow><mn>1</mn><mo>−</mo><mi>p</mi></mrow><mn>2</mn></mfrac><mo>,</mo></mrow><annotation encoding="application/x-tex">d = \frac{1 - p}{2},</annotation></semantics></math>
where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>d</mi><annotation encoding="application/x-tex">d</annotation></semantics></math>
is the distance measure and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>
is the preservation statistic. The resulting distances range between 0
and 1: a distance of 0 corresponds to a preservation score of 1, while a
distance of 1 corresponds to a preservation score of -1.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dist_jk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/convertPresToDist.html">convertPresToDist</a></span><span class="op">(</span><span class="va">pres_stats_jk</span>, <span class="st">"cor_kIM"</span><span class="op">)</span></span>
<span><span class="va">random_dist_jk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/convertPresToDist.html">convertPresToDist</a></span><span class="op">(</span><span class="va">random_pres_stats_jk</span>, <span class="st">"cor_kIM"</span><span class="op">)</span></span></code></pre></div>
<p>The resulting object is a list, where each list element stores the
distance measures of a module (an original or jackknifed module version,
if jackknifing is used) in a data frame format:</p>
<table class="table table" style="font-size: 13px; margin-left: auto; margin-right: auto;">
<caption style="font-size: initial !important;">
<span style="font-size:13px">Distance measures of the jackknifed POU5F1
module version “POU5F1_jk_APOE” for a couple of example replicate
pairs</span>
</caption>
<thead><tr>
<th style="text-align:center;">
regulator
</th>
<th style="text-align:center;">
type
</th>
<th style="text-align:center;">
id
</th>
<th style="text-align:center;">
gene_removed
</th>
<th style="text-align:center;">
replicate1
</th>
<th style="text-align:center;">
replicate2
</th>
<th style="text-align:center;">
species1
</th>
<th style="text-align:center;">
species2
</th>
<th style="text-align:center;">
dist
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
jk
</td>
<td style="text-align:center;">
POU5F1_jk_APOE
</td>
<td style="text-align:center;">
APOE
</td>
<td style="text-align:center;">
H1c1
</td>
<td style="text-align:center;">
H2c1
</td>
<td style="text-align:center;">
human
</td>
<td style="text-align:center;">
human
</td>
<td style="text-align:center;">
0.138
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
jk
</td>
<td style="text-align:center;">
POU5F1_jk_APOE
</td>
<td style="text-align:center;">
APOE
</td>
<td style="text-align:center;">
H1c1
</td>
<td style="text-align:center;">
H3c1
</td>
<td style="text-align:center;">
human
</td>
<td style="text-align:center;">
human
</td>
<td style="text-align:center;">
0.156
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
jk
</td>
<td style="text-align:center;">
POU5F1_jk_APOE
</td>
<td style="text-align:center;">
APOE
</td>
<td style="text-align:center;">
H1c1
</td>
<td style="text-align:center;">
G1c1
</td>
<td style="text-align:center;">
human
</td>
<td style="text-align:center;">
gorilla
</td>
<td style="text-align:center;">
0.167
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
jk
</td>
<td style="text-align:center;">
POU5F1_jk_APOE
</td>
<td style="text-align:center;">
APOE
</td>
<td style="text-align:center;">
H1c1
</td>
<td style="text-align:center;">
G1c2
</td>
<td style="text-align:center;">
human
</td>
<td style="text-align:center;">
gorilla
</td>
<td style="text-align:center;">
0.151
</td>
</tr>
</tbody>
</table>
<p>The distance measures of a (jackknifed or original) module can be
organized into and plotted as a distance matrix. The function
<tt>plotDistMats</tt> can produce such a plot for a single module
(required input: data frame) or several modules together (required
input: list of data frames), for example it can be used to visualize the
distance matrices of the 6 example modules (the original module
versions) side by side:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dist_examples</span> <span class="op">&lt;-</span> <span class="va">dist_jk</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">example_modules</span>, <span class="st">"_orig"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">dist_examples</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">example_modules</span></span>
<span><span class="fu"><a href="../reference/plotDistMats.html">plotDistMats</a></span><span class="op">(</span><span class="va">dist_examples</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="dist_mat_examples.png" alt="The distance matrices of 6 example modules." width="100%"><p class="caption">
The distance matrices of 6 example modules.
</p>
</div>
<p>It is apparent based on these 6 examples that the distance patterns
vary quite a bit from module to module. Some modules (e.g. FEZF2) have
in general low distance values which points to a high degree of
preservation, while other modules (e.g. NEUROD4) tend to have higher
distance values which points to a worse preservation. While for most
modules the within-species distances are overall lower than the
cross-species distances (this was a selection criteria for the
preservation statistic after all), the magnitude of this difference is
not constant: in case of the SALL4 module the within-species and
cross-species scores span by and large a similar range, while in case of
the POU5F1 module the ape-cynomolgus cross-species scores are
drastically higher than the rest of the scores.</p>
</div>
<div class="section level3">
<h3 id="tree-reconstruction">Tree reconstruction<a class="anchor" aria-label="anchor" href="#tree-reconstruction"></a>
</h3>
<p>To be able to quantitatively compare these different patterns of
preservation, the distance measures are summarized as trees in the next
step of the workflow. Each of the resulting trees stands for a single
module (an original or jackknifed module version, if jackknifing is
used): the tips represent the replicates and the branch lengths
represent the dissimilarity of module topology between the networks of 2
replicates.</p>
<p>The trees are reconstructed using the neighbor-joining algorithm
implemented by <tt>ape:nj</tt>. The function <tt>reconstructTrees</tt>
is a wrapper around this, tailored for the specific purpose:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trees_jk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/reconstructTrees.html">reconstructTrees</a></span><span class="op">(</span><span class="va">dist_jk</span><span class="op">)</span></span>
<span><span class="va">random_trees_jk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/reconstructTrees.html">reconstructTrees</a></span><span class="op">(</span><span class="va">random_dist_jk</span><span class="op">)</span></span></code></pre></div>
<p>The output is a list of <tt>phylo</tt> objects per (jackknifed or
original) module. The function <tt>plotTrees</tt> can plot one or more
of these module trees analogously to <tt>plotDistMats</tt>. Here we
focus on the trees of the 6 example modules again:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spec_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>human <span class="op">=</span> <span class="st">"#4DAF4A"</span>, gorilla <span class="op">=</span> <span class="st">"#377EB8"</span>, cynomolgus <span class="op">=</span> <span class="st">"#9a1ebd"</span><span class="op">)</span></span>
<span><span class="va">tree_examples</span> <span class="op">&lt;-</span> <span class="va">trees_jk</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">example_modules</span>, <span class="st">"_orig"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">tree_examples</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">example_modules</span></span>
<span><span class="fu"><a href="../reference/plotTrees.html">plotTrees</a></span><span class="op">(</span><span class="va">tree_examples</span>, species_colors <span class="op">=</span> <span class="va">spec_colors</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="tree_examples.png" alt="The tree reconstructions of 6 example modules." width="100%"><p class="caption">
The tree reconstructions of 6 example modules.
</p>
</div>
<p>In line with the differences among the distance matrices, the trees
also differ from module to module. FEZF2, the module with generally low
distance measures (i.e. high preservation scores), tends to have short
branch lengths in its tree, while NEUROD4, the module with generally
high distance measures (i.e. low preservation scores), has a large tree
with long branches. In case of the SALL4 module that had similar
within-species and cross-species distance measures, the replicates of
different species are intermixed within the tree, while in case of the
POU5F1 module that has much higher ape-cynomolgus distances than ape-ape
or cynomolgus-cynomolgus distances, the ape and cynomolgus replicates
are clearly separated within the tree and connected by a long internal
branch.</p>
</div>
<div class="section level3">
<h3 id="calculating-tree-based-statistics">Calculating tree-based statistics<a class="anchor" aria-label="anchor" href="#calculating-tree-based-statistics"></a>
</h3>
<p>To quantify these tell-tale characteristics of the module trees,
various tree-based statistics can be calculated:</p>
<ul>
<li><p><strong>Total tree length</strong>: The total length of all
branches in the tree; measures module variability both within and across
species.</p></li>
<li><p><strong>Diversity of a species</strong>: The total length of the
branches connecting the replicates of the given species to each other;
measures module variability within this particular species.</p></li>
<li><p><strong>Within-species diversity</strong>: The sum of the
diversity values across all species; measures module variability within
species in general.</p></li>
<li><p><strong>Monophyleticity of a species</strong>: Indicates whether
the tree is monophyletic for the replicates of the given species. Only
if a module tree is monophyletic for a species of interest can the
module be tested for divergence between this species and all
others.</p></li>
<li><p><strong>Branch length of a species</strong>: The length of the
internal branch that connects the subtree of the replicates belonging to
the given species and the subtree of all other replicates; measures
module variability between the species of interest and all others.
Undefined if the tree is not monophyletic for the given
species.</p></li>
</ul>
<div class="figure" style="text-align: center">
<img src="tree_based_stats.png" alt="Tree-based statistics to characterize the cross-species conservation of network modules." width="100%"><p class="caption">
Tree-based statistics to characterize the cross-species conservation of
network modules.
</p>
</div>
<p>These statistics are all implemented in the function
<tt>calculateTreeStats</tt>:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tree_stats_jk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateTreeStats.html">calculateTreeStats</a></span><span class="op">(</span><span class="va">trees_jk</span><span class="op">)</span></span>
<span><span class="va">random_tree_stats_jk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateTreeStats.html">calculateTreeStats</a></span><span class="op">(</span><span class="va">random_trees_jk</span><span class="op">)</span></span></code></pre></div>
<table class="table table" style="font-size: 13px; margin-left: auto; margin-right: auto;">
<caption style="font-size: initial !important;">
<span style="font-size:13px">Tree-based statistics of the POU5F1 module
and some of its jackknifed module versions</span>
</caption>
<thead><tr>
<th style="text-align:center;">
regulator
</th>
<th style="text-align:center;">
type
</th>
<th style="text-align:center;">
id
</th>
<th style="text-align:center;">
gene_ removed
</th>
<th style="text-align:center;">
total_ tree_length
</th>
<th style="text-align:center;">
within_species_ diversity
</th>
<th style="text-align:center;">
human_ diversity
</th>
<th style="text-align:center;">
gorilla_ diversity
</th>
<th style="text-align:center;">
cynomolgus_ diversity
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
orig
</td>
<td style="text-align:center;">
POU5F1_orig
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
1.038
</td>
<td style="text-align:center;">
0.728
</td>
<td style="text-align:center;">
0.221
</td>
<td style="text-align:center;">
0.034
</td>
<td style="text-align:center;">
0.473
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
jk
</td>
<td style="text-align:center;">
POU5F1_jk_AP1S2
</td>
<td style="text-align:center;">
AP1S2
</td>
<td style="text-align:center;">
1.042
</td>
<td style="text-align:center;">
0.724
</td>
<td style="text-align:center;">
0.220
</td>
<td style="text-align:center;">
0.034
</td>
<td style="text-align:center;">
0.470
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
jk
</td>
<td style="text-align:center;">
POU5F1_jk_APOE
</td>
<td style="text-align:center;">
APOE
</td>
<td style="text-align:center;">
1.036
</td>
<td style="text-align:center;">
0.734
</td>
<td style="text-align:center;">
0.221
</td>
<td style="text-align:center;">
0.036
</td>
<td style="text-align:center;">
0.476
</td>
</tr>
<tr>
<td style="text-align:center;">
POU5F1
</td>
<td style="text-align:center;">
jk
</td>
<td style="text-align:center;">
POU5F1_jk_CD24
</td>
<td style="text-align:center;">
CD24
</td>
<td style="text-align:center;">
1.038
</td>
<td style="text-align:center;">
0.733
</td>
<td style="text-align:center;">
0.225
</td>
<td style="text-align:center;">
0.039
</td>
<td style="text-align:center;">
0.469
</td>
</tr>
</tbody>
</table>
<table class="table table" style="font-size: 13px; margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:center;">
human_ monophyl
</th>
<th style="text-align:center;">
gorilla_ monophyl
</th>
<th style="text-align:center;">
cynomolgus_ monophyl
</th>
<th style="text-align:center;">
human_ subtree_length
</th>
<th style="text-align:center;">
gorilla_ subtree_length
</th>
<th style="text-align:center;">
cynomolgus_ subtree_length
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:center;">
TRUE
</td>
<td style="text-align:center;">
TRUE
</td>
<td style="text-align:center;">
TRUE
</td>
<td style="text-align:center;">
0.258
</td>
<td style="text-align:center;">
0.120
</td>
<td style="text-align:center;">
0.661
</td>
</tr>
<tr>
<td style="text-align:center;">
TRUE
</td>
<td style="text-align:center;">
TRUE
</td>
<td style="text-align:center;">
TRUE
</td>
<td style="text-align:center;">
0.257
</td>
<td style="text-align:center;">
0.123
</td>
<td style="text-align:center;">
0.662
</td>
</tr>
<tr>
<td style="text-align:center;">
TRUE
</td>
<td style="text-align:center;">
TRUE
</td>
<td style="text-align:center;">
TRUE
</td>
<td style="text-align:center;">
0.256
</td>
<td style="text-align:center;">
0.113
</td>
<td style="text-align:center;">
0.666
</td>
</tr>
<tr>
<td style="text-align:center;">
TRUE
</td>
<td style="text-align:center;">
TRUE
</td>
<td style="text-align:center;">
TRUE
</td>
<td style="text-align:center;">
0.261
</td>
<td style="text-align:center;">
0.128
</td>
<td style="text-align:center;">
0.649
</td>
</tr>
</tbody>
</table>
<p>Similarly to the preservation statistics, the tree-based statistics
can also be summarized over all jackknifed version of a module to obtain
an overall estimate along with a confidence interval:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tree_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarizeJackknifeStats.html">summarizeJackknifeStats</a></span><span class="op">(</span><span class="va">tree_stats_jk</span>, </span>
<span>                                      <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"total_tree_length"</span>, </span>
<span>                                        <span class="st">"within_species_diversity"</span>, </span>
<span>                                        <span class="st">"human_branch_length"</span>, </span>
<span>                                        <span class="st">"human_diversity"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">random_tree_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarizeJackknifeStats.html">summarizeJackknifeStats</a></span><span class="op">(</span><span class="va">random_tree_stats_jk</span>, </span>
<span>                                             <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"total_tree_length"</span>, </span>
<span>                                               <span class="st">"within_species_diversity"</span>, </span>
<span>                                               <span class="st">"human_branch_length"</span>, </span>
<span>                                               <span class="st">"human_diversity"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>To confirm whether the tree-based statistics are well-suited for
further analysis, the contrast between the actual and random modules can
be used again for sanity checking. Most tree-based statistics
(especially the ones measuring diversity) are expected to be lower for
the actual modules than for the random modules. For the statistics of
interest, the distributions can be plotted using the function
<tt>plotTreeStatDistributions</tt>:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotTreeStatDistributions.html">plotTreeStatDistributions</a></span><span class="op">(</span><span class="va">tree_stats</span>, </span>
<span>                          <span class="va">random_tree_stats</span>, </span>
<span>                          <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"within_species_diversity"</span>, </span>
<span>                            <span class="st">"total_tree_length"</span>, </span>
<span>                            <span class="st">"human_diversity"</span>, </span>
<span>                            <span class="st">"human_branch_length"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="tree_stat_distributions.png" alt="Distribution of the tree-based statistics for the actual and random modules." width="60%"><p class="caption">
Distribution of the tree-based statistics for the actual and random
modules.
</p>
</div>
<p>For the first 3 statistics (within-species diversity, total tree
length and human diversity), the distributions of the actual and random
modules differ as expected. The 4th statistic (human branch length) is
only defined for modules that are monophyletic for the human replicates,
that is why it could be calculated only for 57 actual modules and 1
random module. Thus the distributions cannot be compared, but finding
fewer monophyletic random modules goes again in the same direction:
higher diversity (lower preservation) among the human replicates in case
of the random modules than in case of the actual modules.</p>
<p><br><br></p>
</div>
</div>
<div class="section level2">
<h2 id="quantifying-cross-species-conservation-of-network-modules">Quantifying cross-species conservation of network modules<a class="anchor" aria-label="anchor" href="#quantifying-cross-species-conservation-of-network-modules"></a>
</h2>
<p>After having calculated first pairwise preservation scores between
replicates and then higher-level tree-based statistics to characterize
how different the topologies of the modules are within and across
species, we are ready to quantify the evolutionary conservation of
modules. But why do we need the within-species differences in the first
place? Why is the extent of cross-species differences not directly the
answer to evolutionary conservation?</p>
<p>There are several sources of module topology differences when we
compare 2 species: there is of course true evolutionary divergence
between the species which is what we are primarily interested in, but
this is convoluted with diversity across individuals of the same
species, technical noise and low statistical power. In the end, if we
find that the topology of a module is different between 2 species, it is
difficult to say to what extent it is actually due to divergence. This
is why the within-species differences also become valuable sources of
information: they provide an estimate of the variability originating
from the other, confounding sources and thus can be used to discern the
signal of real divergence.</p>
<div class="section level3">
<h3 id="overall-conservation-and-divergence">Overall conservation and divergence<a class="anchor" aria-label="anchor" href="#overall-conservation-and-divergence"></a>
</h3>
<p>To apply this concept for the quantification of overall conservation
across all species, we used 2 tree-based statistics: within-species
diversity and total tree length. The two metrics are based on the same
trees and calculated using in part the same branches, so they are
clearly not independent. Their general relationship can be captures by
linear regression, that in biological terms describes which part of the
total module variability is due to sources other than cross-species
divergence (e.g. due to diversity, noise and power).</p>
<p>Where a module is located along the regression line informs us about
detection robustness. Modules that have both low within-species
diversity and low total tree length are well-preserved both within and
across species, meaning that these modules could be robustly detected in
all replicates, whereas modules, for which both metrics are high, are
poorly preserved not just across but also within species, indicating a
high detection uncertainty (“wobbliness”).</p>
<p>But more importantly, the residuals of the modules in the regression
model informs us about their cross-species conservation: the outlier
modules that do not follow the general trend are the most conserved and
diverged ones. The modules that are located above the trend line have a
higher total tree length than expected based on their within-species
diversity, meaning that they tend to have long internal branches
connecting the species, which indicates divergence. In contrast, the
modules that fall below the trend line have a lower total tree length
than expected based on their within-species diversity, meaning that the
species are not well-separated within the tree but rather mixed amog
each other which hints towards conservation.</p>
<div class="figure" style="text-align: center">
<img src="within_total_concept.png" alt="The concept of using tree-based statistics to puzzle apart conserved, diverged, robust and wobbly modules." width="38%"><p class="caption">
The concept of using tree-based statistics to puzzle apart conserved,
diverged, robust and wobbly modules.
</p>
</div>
<div class="section level4">
<h4 id="filtering-the-module-trees">Filtering the module trees<a class="anchor" aria-label="anchor" href="#filtering-the-module-trees"></a>
</h4>
<p>The general trend between the within-species diversities and total
tree lengths can be visualized using the function
<tt>plotTreeStats</tt>. In the plot below, both the actual and the
random modules are included:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotTreeStats.html">plotTreeStats</a></span><span class="op">(</span><span class="va">tree_stats</span>, </span>
<span>              <span class="va">random_tree_stats</span>, </span>
<span>              <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"within_species_diversity"</span>, <span class="st">"total_tree_length"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="within_species_diversity_total_tree_length.png" alt="The relationship of the total tree lengths and within-species diversities across all actual and random modules." width="75%"><p class="caption">
The relationship of the total tree lengths and within-species
diversities across all actual and random modules.
</p>
</div>
<p>As we have seen before, the within-species diversities and total tree
lengths are in general higher for the random modules than for the actual
modules - thus the random modules cluster separately on the plot at the
“wobbly” end of the spectrum. However, there are still individual actual
modules that have particularly high total tree lengths and
within-species diversities, comparable with those of the random modules.
Since these modules seem to behave like random sets of genes in terms of
their tree properties, it is recommended to remove them for further
analysis.</p>
<p>The function <tt>filterModuleTrees</tt> calculates how probable it is
that the statistic of an actual comes from the distribution of all
actual modules compared
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>p</mi><mrow><mi>a</mi><mi>c</mi><mi>t</mi><mi>u</mi><mi>a</mi><mi>l</mi></mrow></msub><annotation encoding="application/x-tex">p_{actual}</annotation></semantics></math>)
and how probable it is that the statistic comes from the distribution of
all random modules
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>p</mi><mrow><mi>r</mi><mi>a</mi><mi>n</mi><mi>d</mi><mi>o</mi><mi>m</mi></mrow></msub><annotation encoding="application/x-tex">p_{random}</annotation></semantics></math>)
using one-sample one-sided z-tests (expected mean and standard
deviation: the mean and standard deviation of all actual/random module
statistics after outlier removal using the interquartile range method).
If
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>1</mn><mo>−</mo><msub><mi>p</mi><mrow><mi>a</mi><mi>c</mi><mi>t</mi><mi>u</mi><mi>a</mi><mi>l</mi></mrow></msub><mo>&gt;</mo><mn>10</mn><mo>⋅</mo><msub><mi>p</mi><mrow><mi>r</mi><mi>a</mi><mi>n</mi><mi>d</mi><mi>o</mi><mi>m</mi></mrow></msub></mrow><annotation encoding="application/x-tex">1 - p_{actual} &gt; 10⋅p_{random}</annotation></semantics></math>
is fulfilled for both tree statistics, the module is kept, otherwise it
is removed.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tree_stats_filt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterModuleTrees.html">filterModuleTrees</a></span><span class="op">(</span><span class="va">tree_stats</span>, <span class="va">random_tree_stats</span><span class="op">)</span></span></code></pre></div>
<p>In case of the example dataset, this filtering removes 4 modules
(marked in red):</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotTreeStats.html">plotTreeStats</a></span><span class="op">(</span><span class="va">tree_stats</span>, </span>
<span>              <span class="va">random_tree_stats</span>, </span>
<span>              <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"within_species_diversity"</span>, <span class="st">"total_tree_length"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>data <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter-joins.html" class="external-link">anti_join</a></span><span class="op">(</span><span class="va">tree_stats</span>, <span class="va">tree_stats_filt</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"red3"</span>, size <span class="op">=</span> <span class="fl">1.2</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="within_species_diversity_total_tree_length_filt.png" alt="The relationship of the total tree lengths and within-species diversities across all actual and random modules, with the removed actual modules marked." width="75%"><p class="caption">
The relationship of the total tree lengths and within-species
diversities across all actual and random modules, with the removed
actual modules marked.
</p>
</div>
</div>
<div class="section level4">
<h4 id="finding-conserved-and-diverged-modules">Finding conserved and diverged modules<a class="anchor" aria-label="anchor" href="#finding-conserved-and-diverged-modules"></a>
</h4>
<p>After removing these “wobbly” modules, cross-species conservation can
be quantified using a linear regression between the total tree length
and within-species branch lengths.</p>
<p>The function <tt>fitTreeStatsLm</tt> fits a regression model between
the 2 statistics and the function <tt>findConservedDivergedModules</tt>
identifies conserved and diverged modules by comparing each module to
the 95% prediction interval of the fit. A module is considered diverged
if it has a higher total tree length than the upper boundary of the
prediction interval, while a module is considered conserved if it has a
lower total tree length than the lower boundary of the prediction
interval.</p>
<p>The degree of conservation/divergence can be quantitatively compared
between the modules categorized as conserved/diverged. We implemented 2
measures: 1) the residual, which is the absolute difference between the
observed and expected total tree lengths, and 2) the t-score, which is
the residual normalized by the standard error of the total tree length
prediction at a given within-species diversity value.</p>
<p>The linear regression can be weighted by the error of the data points
derived from jackknifing. If this option is chosen (<tt>weighted_lm</tt>
is set to TRUE), the weight of a module in the regression is defined as
inversely proportional to the total tree length variance across all of
its jackknifed module versions.</p>
<p>In case of the example dataset, we used a weighted linear
regression:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lm_overall</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitTreeStatsLm.html">fitTreeStatsLm</a></span><span class="op">(</span><span class="va">tree_stats_filt</span>, </span>
<span>                             focus <span class="op">=</span> <span class="st">"overall"</span>,</span>
<span>                             weighted_lm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">module_conservation_overall</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findConservedDivergedModules.html">findConservedDivergedModules</a></span><span class="op">(</span><span class="va">tree_stats_filt</span>, </span>
<span>                                                            <span class="va">lm_overall</span><span class="op">)</span></span></code></pre></div>
<table class="table table" style="font-size: 13px; margin-left: auto; margin-right: auto;">
<caption style="font-size: initial !important;">
<span style="font-size:13px">Cross-species conservation measures for 6
example modules</span>
</caption>
<thead><tr>
<th style="text-align:center;">
focus
</th>
<th style="text-align:center;">
regulator
</th>
<th style="text-align:center;">
module_ size
</th>
<th style="text-align:center;">
total_ tree_length
</th>
<th style="text-align:center;">
lwr_total_ tree_length
</th>
<th style="text-align:center;">
upr_total_ tree_length
</th>
<th style="text-align:center;">
within_species_ diversity
</th>
<th style="text-align:center;">
lwr_within_ species_diversity
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:center;width: 2.3cm; ">
overall
</td>
<td style="text-align:center;width: 2.3cm; ">
NANOG
</td>
<td style="text-align:center;width: 2.3cm; ">
42
</td>
<td style="text-align:center;width: 2.3cm; ">
0.599
</td>
<td style="text-align:center;width: 2.3cm; ">
0.596
</td>
<td style="text-align:center;width: 2.3cm; ">
0.604
</td>
<td style="text-align:center;width: 2.3cm; ">
0.457
</td>
<td style="text-align:center;width: 2.3cm; ">
0.456
</td>
</tr>
<tr>
<td style="text-align:center;width: 2.3cm; ">
overall
</td>
<td style="text-align:center;width: 2.3cm; ">
POU5F1
</td>
<td style="text-align:center;width: 2.3cm; ">
46
</td>
<td style="text-align:center;width: 2.3cm; ">
1.036
</td>
<td style="text-align:center;width: 2.3cm; ">
1.033
</td>
<td style="text-align:center;width: 2.3cm; ">
1.042
</td>
<td style="text-align:center;width: 2.3cm; ">
0.730
</td>
<td style="text-align:center;width: 2.3cm; ">
0.726
</td>
</tr>
<tr>
<td style="text-align:center;width: 2.3cm; ">
overall
</td>
<td style="text-align:center;width: 2.3cm; ">
SALL4
</td>
<td style="text-align:center;width: 2.3cm; ">
40
</td>
<td style="text-align:center;width: 2.3cm; ">
1.130
</td>
<td style="text-align:center;width: 2.3cm; ">
1.120
</td>
<td style="text-align:center;width: 2.3cm; ">
1.143
</td>
<td style="text-align:center;width: 2.3cm; ">
1.191
</td>
<td style="text-align:center;width: 2.3cm; ">
1.173
</td>
</tr>
<tr>
<td style="text-align:center;width: 2.3cm; ">
overall
</td>
<td style="text-align:center;width: 2.3cm; ">
NEUROD4
</td>
<td style="text-align:center;width: 2.3cm; ">
52
</td>
<td style="text-align:center;width: 2.3cm; ">
1.208
</td>
<td style="text-align:center;width: 2.3cm; ">
1.200
</td>
<td style="text-align:center;width: 2.3cm; ">
1.214
</td>
<td style="text-align:center;width: 2.3cm; ">
1.254
</td>
<td style="text-align:center;width: 2.3cm; ">
1.240
</td>
</tr>
<tr>
<td style="text-align:center;width: 2.3cm; ">
overall
</td>
<td style="text-align:center;width: 2.3cm; ">
PAX6
</td>
<td style="text-align:center;width: 2.3cm; ">
48
</td>
<td style="text-align:center;width: 2.3cm; ">
0.798
</td>
<td style="text-align:center;width: 2.3cm; ">
0.794
</td>
<td style="text-align:center;width: 2.3cm; ">
0.804
</td>
<td style="text-align:center;width: 2.3cm; ">
0.643
</td>
<td style="text-align:center;width: 2.3cm; ">
0.639
</td>
</tr>
<tr>
<td style="text-align:center;width: 2.3cm; ">
overall
</td>
<td style="text-align:center;width: 2.3cm; ">
FEZF2
</td>
<td style="text-align:center;width: 2.3cm; ">
40
</td>
<td style="text-align:center;width: 2.3cm; ">
0.560
</td>
<td style="text-align:center;width: 2.3cm; ">
0.554
</td>
<td style="text-align:center;width: 2.3cm; ">
0.567
</td>
<td style="text-align:center;width: 2.3cm; ">
0.434
</td>
<td style="text-align:center;width: 2.3cm; ">
0.426
</td>
</tr>
</tbody>
</table>
<table class="table table" style="font-size: 13px; margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:center;">
upr_within_ species_diversity
</th>
<th style="text-align:center;">
fit
</th>
<th style="text-align:center;">
lwr_fit
</th>
<th style="text-align:center;">
upr_fit
</th>
<th style="text-align:center;">
residual
</th>
<th style="text-align:center;">
weight
</th>
<th style="text-align:center;">
t_score
</th>
<th style="text-align:center;">
conservation
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:center;width: 2.3cm; ">
0.461
</td>
<td style="text-align:center;width: 2.3cm; ">
0.654
</td>
<td style="text-align:center;width: 2.3cm; ">
0.571
</td>
<td style="text-align:center;width: 2.3cm; ">
0.737
</td>
<td style="text-align:center;width: 2.3cm; ">
-0.055
</td>
<td style="text-align:center;width: 2.3cm; ">
0.003
</td>
<td style="text-align:center;width: 2.3cm; ">
-1.296
</td>
<td style="text-align:center;width: 2.3cm; ">
not_significant
</td>
</tr>
<tr>
<td style="text-align:center;width: 2.3cm; ">
0.733
</td>
<td style="text-align:center;width: 2.3cm; ">
0.881
</td>
<td style="text-align:center;width: 2.3cm; ">
0.815
</td>
<td style="text-align:center;width: 2.3cm; ">
0.946
</td>
<td style="text-align:center;width: 2.3cm; ">
0.156
</td>
<td style="text-align:center;width: 2.3cm; ">
0.004
</td>
<td style="text-align:center;width: 2.3cm; ">
4.650
</td>
<td style="text-align:center;width: 2.3cm; ">
diverged
</td>
</tr>
<tr>
<td style="text-align:center;width: 2.3cm; ">
1.204
</td>
<td style="text-align:center;width: 2.3cm; ">
1.264
</td>
<td style="text-align:center;width: 2.3cm; ">
1.092
</td>
<td style="text-align:center;width: 2.3cm; ">
1.437
</td>
<td style="text-align:center;width: 2.3cm; ">
-0.134
</td>
<td style="text-align:center;width: 2.3cm; ">
0.001
</td>
<td style="text-align:center;width: 2.3cm; ">
-1.533
</td>
<td style="text-align:center;width: 2.3cm; ">
not_significant
</td>
</tr>
<tr>
<td style="text-align:center;width: 2.3cm; ">
1.262
</td>
<td style="text-align:center;width: 2.3cm; ">
1.317
</td>
<td style="text-align:center;width: 2.3cm; ">
1.178
</td>
<td style="text-align:center;width: 2.3cm; ">
1.456
</td>
<td style="text-align:center;width: 2.3cm; ">
-0.109
</td>
<td style="text-align:center;width: 2.3cm; ">
0.001
</td>
<td style="text-align:center;width: 2.3cm; ">
-1.545
</td>
<td style="text-align:center;width: 2.3cm; ">
not_significant
</td>
</tr>
<tr>
<td style="text-align:center;width: 2.3cm; ">
0.647
</td>
<td style="text-align:center;width: 2.3cm; ">
0.808
</td>
<td style="text-align:center;width: 2.3cm; ">
0.733
</td>
<td style="text-align:center;width: 2.3cm; ">
0.884
</td>
<td style="text-align:center;width: 2.3cm; ">
-0.011
</td>
<td style="text-align:center;width: 2.3cm; ">
0.003
</td>
<td style="text-align:center;width: 2.3cm; ">
-0.274
</td>
<td style="text-align:center;width: 2.3cm; ">
not_significant
</td>
</tr>
<tr>
<td style="text-align:center;width: 2.3cm; ">
0.442
</td>
<td style="text-align:center;width: 2.3cm; ">
0.635
</td>
<td style="text-align:center;width: 2.3cm; ">
0.523
</td>
<td style="text-align:center;width: 2.3cm; ">
0.746
</td>
<td style="text-align:center;width: 2.3cm; ">
-0.075
</td>
<td style="text-align:center;width: 2.3cm; ">
0.001
</td>
<td style="text-align:center;width: 2.3cm; ">
-1.324
</td>
<td style="text-align:center;width: 2.3cm; ">
not_significant
</td>
</tr>
</tbody>
</table>
<p>Using this approach, we identified 21 conserved and 23 diverged
modules:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">module_conservation_overall</span><span class="op">$</span><span class="va">conservation</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;       conserved        diverged not_significant </span></span>
<span><span class="co">#&gt;              21              23             788</span></span></code></pre></div>
<p>The function <tt>plotConservedDivergedModules</tt> helps to visualize
the regression line, its 95% prediction interval (grey area), and the
conserved and diverged modules (colored green and red, respectively). It
also labels the top N most conserved and most diverged modules based on
either the residuals or the t-scores. In the plot below, the top 5 most
conserved and most diverged modules based on residuals are
highlighted:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotConservedDivergedModules.html">plotConservedDivergedModules</a></span><span class="op">(</span><span class="va">module_conservation_overall</span>,</span>
<span>                             N <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                             rank_by <span class="op">=</span> <span class="st">"residual"</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="module_conservation_overall.png" alt="Weighted linear regression between the total tree length and within-species diversity, with the top 5 most conserved and most diverged modules marked." width="75%"><p class="caption">
Weighted linear regression between the total tree length and
within-species diversity, with the top 5 most conserved and most
diverged modules marked.
</p>
</div>
<p>To illustrate what makes these modules conserved or diverged, their
distance matrices and module trees can also be plotted:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top5_cons_div_modules</span> <span class="op">&lt;-</span> <span class="fu">bind_rows</span><span class="op">(</span><span class="va">module_conservation_overall</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>                                    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">conservation</span> <span class="op">==</span> <span class="st">"conserved"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>                                    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_min</a></span><span class="op">(</span>order_by <span class="op">=</span> <span class="va">residual</span>, n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>,</span>
<span>                                   <span class="va">module_conservation_overall</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>                                    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">conservation</span> <span class="op">==</span> <span class="st">"diverged"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>                                    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_max</a></span><span class="op">(</span>order_by <span class="op">=</span> <span class="va">residual</span>, n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">regulator</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dist</span> <span class="op">&lt;-</span> <span class="va">dist_jk</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">top5_cons_div_modules</span>, <span class="st">"_orig"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">dist</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">top5_cons_div_modules</span></span>
<span><span class="fu"><a href="../reference/plotDistMats.html">plotDistMats</a></span><span class="op">(</span><span class="va">dist</span>, ncol <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="dist_mat_cons_div.png" alt="The distance matrices of the 5 most conserved (top row) and 5 most diverged (bottom row) modules." width="100%"><p class="caption">
The distance matrices of the 5 most conserved (top row) and 5 most
diverged (bottom row) modules.
</p>
</div>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trees</span> <span class="op">&lt;-</span> <span class="va">trees_jk</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">top5_cons_div_modules</span>, <span class="st">"_orig"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">trees</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">top5_cons_div_modules</span></span>
<span><span class="fu"><a href="../reference/plotTrees.html">plotTrees</a></span><span class="op">(</span><span class="va">trees</span>, species_colors <span class="op">=</span> <span class="va">spec_colors</span>, ncol <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="trees_cons_div.png" alt="The tree reconstructions of the 5 most conserved (top row) and 5 most diverged (bottom row) modules." width="100%"><p class="caption">
The tree reconstructions of the 5 most conserved (top row) and 5 most
diverged (bottom row) modules.
</p>
</div>
<p>The contrast between the conserved and diverged modules in terms of
their distance matrices and module trees is clearly visible. The
conserved modules tend to have balanced distance matrices with similarly
low distance measures for all within-species and cross-species
comparisons, and correspondingly, tree reconstructions without any
species separation, with replicates from the same species being as close
to each other as replicates from different species.</p>
<p>In contrast, diverged modules tend to have much higher distance
measures across species than within species for 1 or more species pairs
(in case of the example modules it is always the human-cynomolgus and
gorilla-cynomolgus species pairs). This also results in trees where
these species are strikingly distinct, with short branches within the
species or clades and a long internal branch separating these subtrees
from each other.</p>
</div>
<div class="section level4">
<h4 id="finding-conserved-and-diverged-edges">Finding conserved and diverged edges<a class="anchor" aria-label="anchor" href="#finding-conserved-and-diverged-edges"></a>
</h4>
<p>A divergence score can be calculated not just for a module as a
whole, but also for each edge within a module, which can pinpoint
individual edges or regions of neighboring edges that are particularly
conserved or diverged. This score is calculated based on the different
edge weights of a given edge in the networks of the different
replicates. The edge weights are compared across species using an ANOVA,
and the log<sub>10</sub> of the F-statistic (i.e. between-species
variability / within-species variability) is regarded as the measure of
edge divergence.</p>
<p>These scores can be returned as a data frame using
<tt>calculateEdgeDivergence</tt> or directly plotted using
<tt>plotNetworks</tt> with <tt>color_by</tt> set to “edge_divergence”.
Below we plot the 300 strongest edges within each of the 5 most
conserved and most diverged modules colored by edge divergence:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotNetworks.html">plotNetworks</a></span><span class="op">(</span><span class="va">top5_cons_div_modules</span>, </span>
<span>             <span class="va">pruned_modules</span>, </span>
<span>             <span class="va">consensus_network</span>, </span>
<span>             <span class="va">network_list</span>, </span>
<span>             <span class="va">replicate2species</span>, </span>
<span>             color_by <span class="op">=</span> <span class="st">"edge_divergence"</span>,</span>
<span>             ncol <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="networks_cons_div.png" alt="The 300 strongest connections of the 5 most conserved (top row) and 5 most diverged (bottom row) modules. The thickness of the edges represents the consensus edge weights and the color of the edges represents how different the mean edge weights are across the 3 species." width="100%"><p class="caption">
The 300 strongest connections of the 5 most conserved (top row) and 5
most diverged (bottom row) modules. The thickness of the edges
represents the consensus edge weights and the color of the edges
represents how different the mean edge weights are across the 3 species.
</p>
</div>
<p>Based on the network visualizations, the diverged modules (especially
HMGA1 and POU5F1) tend to have more diverged edges than the conserved
modules, which confirms the previous findings from a different aspect.
The genes that form the most diverged edges within a module can also be
pinned down, for example in case of the POU5F1 module, SCGB3A2 is the
main hub for the diverged edges.</p>
</div>
<div class="section level4">
<h4 id="finding-conserved-and-diverged-target-genes">Finding conserved and diverged target genes<a class="anchor" aria-label="anchor" href="#finding-conserved-and-diverged-target-genes"></a>
</h4>
<p>Genes that contribute the most to the conservation/divergence of a
module can be identified not just using the edge divergence scores, but
also using the tree-based statistics in combination with
jackknifing.</p>
<p>During jackknifing, each member gene of a module is removed and all
statistics are recalculated, including the tree-based statistics (total
tree length and within-species branch length) that inform us about
cross-species conservation. Our working hypothesis is that if removing a
target gene from a diverged module makes the module more conserved, then
that target was responsible for divergence in the original module, and
vice versa, if removing a target gene from a conserved module makes the
module more diverged, then that target was responsible for conservation
in the original module.</p>
<p>To quantify these effects, the function
<tt>findConservedDivergedTargets</tt> takes use of the linear model that
was fitted across all modules between the total tree length and
within-species diversity and that was used to identify the conserved and
diverged modules in the first place. However, in this case it is not the
aggregate statistic for a module as a whole that is compared to the
regression line, but the statistic of each jackknife module version
separately. The corresponding residuals quantify the extent of target
gene conservation: the jackknife version that has the highest residual
is the most diverged, therefore the corresponding target gene is the
most conserved, while the jackknife version that has the lowest residual
is the most conserved, therefore the corresponding target gene is the
most diverged.</p>
<p>It is important to note, that in most cases the
divergence/conservation of a module cannot be contributed to a single
target gene, but it is rather the combined signal of all targets
together. The “most conserved” target gene of a diverged module is still
most likely diverged, just to a lesser extent than the others. That is
why it mainly makes sense to investigate the most diverged targets of
the diverged modules and the most conserved targets of the conserved
modules, not the other way around.</p>
<p>In case of the example dataset, we took a closer look at the target
genes of POU5F1:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">POU5F1_target_conservation</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findConservedDivergedTargets.html">findConservedDivergedTargets</a></span><span class="op">(</span><span class="st">"POU5F1"</span>,</span>
<span>                                                           <span class="va">tree_stats_jk</span>,</span>
<span>                                                           <span class="va">lm_overall</span><span class="op">)</span></span></code></pre></div>
<table class="table table" style="font-size: 13px; margin-left: auto; margin-right: auto;">
<caption style="font-size: initial !important;">
<span style="font-size:13px">Cross-species conservation measures for a
few example target genes of POU5F1</span>
</caption>
<thead><tr>
<th style="text-align:center;">
focus
</th>
<th style="text-align:center;">
regulator
</th>
<th style="text-align:center;">
type
</th>
<th style="text-align:center;">
id
</th>
<th style="text-align:center;">
gene_ removed
</th>
<th style="text-align:center;">
within_species_ diversity
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:center;width: 2cm; ">
overall
</td>
<td style="text-align:center;width: 2.1cm; ">
POU5F1
</td>
<td style="text-align:center;width: 1.7cm; ">
orig
</td>
<td style="text-align:center;width: 3.1cm; ">
POU5F1_orig
</td>
<td style="text-align:center;width: 2.3cm; ">
NA
</td>
<td style="text-align:center;width: 2.6cm; ">
0.728
</td>
</tr>
<tr>
<td style="text-align:center;width: 2cm; ">
overall
</td>
<td style="text-align:center;width: 2.1cm; ">
POU5F1
</td>
<td style="text-align:center;width: 1.7cm; ">
jk
</td>
<td style="text-align:center;width: 3.1cm; ">
POU5F1_jk_AP1S2
</td>
<td style="text-align:center;width: 2.3cm; ">
AP1S2
</td>
<td style="text-align:center;width: 2.6cm; ">
0.724
</td>
</tr>
<tr>
<td style="text-align:center;width: 2cm; ">
overall
</td>
<td style="text-align:center;width: 2.1cm; ">
POU5F1
</td>
<td style="text-align:center;width: 1.7cm; ">
jk
</td>
<td style="text-align:center;width: 3.1cm; ">
POU5F1_jk_APOE
</td>
<td style="text-align:center;width: 2.3cm; ">
APOE
</td>
<td style="text-align:center;width: 2.6cm; ">
0.734
</td>
</tr>
<tr>
<td style="text-align:center;width: 2cm; ">
overall
</td>
<td style="text-align:center;width: 2.1cm; ">
POU5F1
</td>
<td style="text-align:center;width: 1.7cm; ">
jk
</td>
<td style="text-align:center;width: 3.1cm; ">
POU5F1_jk_CD24
</td>
<td style="text-align:center;width: 2.3cm; ">
CD24
</td>
<td style="text-align:center;width: 2.6cm; ">
0.733
</td>
</tr>
</tbody>
</table>
<table class="table table" style="font-size: 13px; margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:center;">
total_ tree_length
</th>
<th style="text-align:center;">
fit
</th>
<th style="text-align:center;">
lwr_fit
</th>
<th style="text-align:center;">
upr_fit
</th>
<th style="text-align:center;">
residual
</th>
<th style="text-align:center;">
contribution
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:center;width: 2.3cm; ">
1.038
</td>
<td style="text-align:center;width: 2.3cm; ">
0.879
</td>
<td style="text-align:center;width: 2.3cm; ">
0.813
</td>
<td style="text-align:center;width: 2.3cm; ">
0.945
</td>
<td style="text-align:center;width: 2.3cm; ">
0.159
</td>
<td style="text-align:center;width: 2.3cm; ">
NA
</td>
</tr>
<tr>
<td style="text-align:center;width: 2.3cm; ">
1.042
</td>
<td style="text-align:center;width: 2.3cm; ">
0.876
</td>
<td style="text-align:center;width: 2.3cm; ">
0.810
</td>
<td style="text-align:center;width: 2.3cm; ">
0.942
</td>
<td style="text-align:center;width: 2.3cm; ">
0.166
</td>
<td style="text-align:center;width: 2.3cm; ">
-0.045
</td>
</tr>
<tr>
<td style="text-align:center;width: 2.3cm; ">
1.036
</td>
<td style="text-align:center;width: 2.3cm; ">
0.884
</td>
<td style="text-align:center;width: 2.3cm; ">
0.818
</td>
<td style="text-align:center;width: 2.3cm; ">
0.949
</td>
<td style="text-align:center;width: 2.3cm; ">
0.152
</td>
<td style="text-align:center;width: 2.3cm; ">
0.041
</td>
</tr>
<tr>
<td style="text-align:center;width: 2.3cm; ">
1.038
</td>
<td style="text-align:center;width: 2.3cm; ">
0.883
</td>
<td style="text-align:center;width: 2.3cm; ">
0.817
</td>
<td style="text-align:center;width: 2.3cm; ">
0.949
</td>
<td style="text-align:center;width: 2.3cm; ">
0.155
</td>
<td style="text-align:center;width: 2.3cm; ">
0.023
</td>
</tr>
</tbody>
</table>
<p>The original and jackknifed versions of a module can be plotted along
the regression line with the help of the function
<tt>plotConservedDivergedTargets</tt>. The prediction interval of the
regression line is shown as a grey area and assumed to have a constant
width for a given module. This is how jackknifing affects the position
compared to the regression line in case of the POU5F1 module:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotConservedDivergedTargets.html">plotConservedDivergedTargets</a></span><span class="op">(</span><span class="va">POU5F1_target_conservation</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="POU5F1_target_conservation.png" alt="The top 5 most diverged target genes of POU5F1 identified using tree-based statistics in combination with jackknifing." width="75%"><p class="caption">
The top 5 most diverged target genes of POU5F1 identified using
tree-based statistics in combination with jackknifing.
</p>
</div>
<p>All jackknifed versions are located way outside the prediction
interval, meaning that the whole module exhibits a strong signal of
divergence that cannot be explained by any of the target genes alone.
The most diverged target with the lowest residual is SCGB3A2, in line
with the findings based on edge divergence.</p>
<p>To investigate whether the network divergence/conservation also
translates into expression divergence/conservation, the expression
profiles of the regulator and its targets can be plotted per species
using <tt>plotExprAlongPseudotime</tt> (for pseudotime trajectories) or
<tt>plotExprPerCellType</tt> (for distinct cell types). Below, we plot
the expression of POU5F1 and its 5 most diverged target genes across the
pseudotime range of the neural differentiation process:</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top5_diverged_POU5F1_targets</span> <span class="op">&lt;-</span> <span class="va">POU5F1_target_conservation</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">slice_min</span><span class="op">(</span>order_by <span class="op">=</span> <span class="va">residual</span>, n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">pull</span><span class="op">(</span><span class="va">gene_removed</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ct_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Pluripotent_Cells <span class="op">=</span> <span class="st">"#86C1E6"</span>, Early_Ectoderm <span class="op">=</span> <span class="st">"#F4AB62"</span>, Neurons <span class="op">=</span> <span class="st">"#CA6102"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotExprAlongPseudotime.html">plotExprAlongPseudotime</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"POU5F1"</span>, <span class="va">top5_diverged_POU5F1_targets</span><span class="op">)</span>, </span>
<span>                        <span class="va">sce</span>, </span>
<span>                        species_colors <span class="op">=</span> <span class="va">spec_colors</span>, </span>
<span>                        cell_type_colors <span class="op">=</span> <span class="va">ct_colors</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="POU5F1_target_expression.png" alt="The expression profiles of POU5F1 and its top 5 most diverged target genes across pseudotime trajectory per species." width="60%"><p class="caption">
The expression profiles of POU5F1 and its top 5 most diverged target
genes across pseudotime trajectory per species.
</p>
</div>
<p>Several target genes of POU5F1 that are diverged in terms of network
topology (SCGB3A2, SPP1 and FTL) also show different expression patterns
across species. Such differences cannot be observed for POU5F1, the
regulator itself - a hint towards cis-regulatory changes. Indeed, after
a more thorough investigation we found that both SCGB3A2 and SPP1 are
known to be located close to LTR7 elements in the human genome <span class="citation">[25, 26]</span>, an evolutionary young family of
transposable elements that often acquire regulatory functions. The LTR7
elements near these 2 genes appeared along the ape lineage and are bound
by POU5F1 <span class="citation">[27]</span>, so they could have
potentially contributed to the rewiring of the POU5F1 module.</p>
</div>
</div>
<div class="section level3">
<h3 id="lineage-specific-divergence">Lineage-specific divergence<a class="anchor" aria-label="anchor" href="#lineage-specific-divergence"></a>
</h3>
<p>The approach described above characterizes the overall degree of
module conservation - all species within a tree are treated the same.
This is the most suitable approach for finding conserved modules and for
finding the highest signal of divergence irrespective of phylogeny. In
case of the example dataset, most of the diverged modules detected this
way are diverged between cynomolgus macaques and great apes (rather than
within the great apes) simply because longer phylogenetic distances
generally give rise to bigger differences.</p>
<p>However, if the focus of interest is divergence on a particular
lineage, for example on the human lineage, a different approach needs to
be applied. In contrast to the overall test of conservation where there
are no constraints in terms of tree structure, here it makes sense to
restrict the testing only to module trees that are monophyletic for the
human replicates. This step pre-selects modules that fulfill a minimal
citeria of lineage-specific divergence and makes it possible to compare
a more specific tree characteristic, namely the human branch length. To
again correct for the unwanted sources of variation, this branch length
is contrasted against variation within the lineage, i.e. the human
diversity. An analogous approach can be applied to find diverged modules
between any species and the rest of the tree as well.</p>
<div class="section level4">
<h4 id="finding-conserved-and-diverged-modules-1">Finding conserved and diverged modules<a class="anchor" aria-label="anchor" href="#finding-conserved-and-diverged-modules-1"></a>
</h4>
<p>The modules that are diverged on the human lineage can be defined
using a (weighted) linear regression between the human branch length and
the human diversity, implemented in <tt>fitTreeStatsLm</tt> with
<tt>focus</tt> set to “human” and <tt>findConservedDivergedModules</tt>.
Similarly to the overall conservation and divergence, the 95% prediction
interval of the regression line is calculated and the modules that have
a higher human branch length than the upper bound of the 95% prediction
interval are considered diverged. It is important to note that the
modules that fall below the lower bound of the 95% prediction interval
cannot be interpreted as conserved in this case, because 1) the
human-monophyletic module trees have already been pre-selected, and 2)
conservation is per definition not a lineage-specific feature.</p>
<p>Below, this approach is applied to the example dataset:</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lm_human</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitTreeStatsLm.html">fitTreeStatsLm</a></span><span class="op">(</span><span class="va">tree_stats_filt</span>, </span>
<span>                           focus <span class="op">=</span> <span class="st">"human"</span>,</span>
<span>                           weighted_lm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">module_conservation_human</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findConservedDivergedModules.html">findConservedDivergedModules</a></span><span class="op">(</span><span class="va">tree_stats_filt</span>, </span>
<span>                                                          <span class="va">lm_human</span><span class="op">)</span></span></code></pre></div>
<table class="table table" style="font-size: 13px; margin-left: auto; margin-right: auto;">
<caption style="font-size: initial !important;">
<span style="font-size:13px">Measures of divergence on the human lineage
for 6 example modules</span>
</caption>
<thead><tr>
<th style="text-align:center;">
focus
</th>
<th style="text-align:center;">
regulator
</th>
<th style="text-align:center;">
module_ size
</th>
<th style="text-align:center;">
human_ subtree_length
</th>
<th style="text-align:center;">
lwr_human_subtree_ length
</th>
<th style="text-align:center;">
upr_human_subtree_ length
</th>
<th style="text-align:center;">
human_ diversity
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:center;width: 2.5cm; ">
human
</td>
<td style="text-align:center;width: 2.5cm; ">
AR
</td>
<td style="text-align:center;width: 2.5cm; ">
39
</td>
<td style="text-align:center;width: 2.5cm; ">
0.362
</td>
<td style="text-align:center;width: 2.5cm; ">
0.359
</td>
<td style="text-align:center;width: 2.5cm; ">
0.367
</td>
<td style="text-align:center;width: 2.5cm; ">
0.335
</td>
</tr>
<tr>
<td style="text-align:center;width: 2.5cm; ">
human
</td>
<td style="text-align:center;width: 2.5cm; ">
ARID1B
</td>
<td style="text-align:center;width: 2.5cm; ">
43
</td>
<td style="text-align:center;width: 2.5cm; ">
0.426
</td>
<td style="text-align:center;width: 2.5cm; ">
0.423
</td>
<td style="text-align:center;width: 2.5cm; ">
0.434
</td>
<td style="text-align:center;width: 2.5cm; ">
0.338
</td>
</tr>
<tr>
<td style="text-align:center;width: 2.5cm; ">
human
</td>
<td style="text-align:center;width: 2.5cm; ">
ARID2
</td>
<td style="text-align:center;width: 2.5cm; ">
48
</td>
<td style="text-align:center;width: 2.5cm; ">
0.330
</td>
<td style="text-align:center;width: 2.5cm; ">
0.324
</td>
<td style="text-align:center;width: 2.5cm; ">
0.335
</td>
<td style="text-align:center;width: 2.5cm; ">
0.304
</td>
</tr>
<tr>
<td style="text-align:center;width: 2.5cm; ">
human
</td>
<td style="text-align:center;width: 2.5cm; ">
BNC2
</td>
<td style="text-align:center;width: 2.5cm; ">
49
</td>
<td style="text-align:center;width: 2.5cm; ">
0.319
</td>
<td style="text-align:center;width: 2.5cm; ">
0.316
</td>
<td style="text-align:center;width: 2.5cm; ">
0.324
</td>
<td style="text-align:center;width: 2.5cm; ">
0.286
</td>
</tr>
<tr>
<td style="text-align:center;width: 2.5cm; ">
human
</td>
<td style="text-align:center;width: 2.5cm; ">
CREB3
</td>
<td style="text-align:center;width: 2.5cm; ">
49
</td>
<td style="text-align:center;width: 2.5cm; ">
0.352
</td>
<td style="text-align:center;width: 2.5cm; ">
0.349
</td>
<td style="text-align:center;width: 2.5cm; ">
0.354
</td>
<td style="text-align:center;width: 2.5cm; ">
0.313
</td>
</tr>
<tr>
<td style="text-align:center;width: 2.5cm; ">
human
</td>
<td style="text-align:center;width: 2.5cm; ">
E2F3
</td>
<td style="text-align:center;width: 2.5cm; ">
46
</td>
<td style="text-align:center;width: 2.5cm; ">
0.390
</td>
<td style="text-align:center;width: 2.5cm; ">
0.385
</td>
<td style="text-align:center;width: 2.5cm; ">
0.395
</td>
<td style="text-align:center;width: 2.5cm; ">
0.355
</td>
</tr>
</tbody>
</table>
<table class="table table" style="font-size: 13px; margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:center;">
lwr_human_ diversity
</th>
<th style="text-align:center;">
upr_human_ diversity
</th>
<th style="text-align:center;">
fit
</th>
<th style="text-align:center;">
lwr_fit
</th>
<th style="text-align:center;">
upr_fit
</th>
<th style="text-align:center;">
residual
</th>
<th style="text-align:center;">
weight
</th>
<th style="text-align:center;">
t_score
</th>
<th style="text-align:center;">
conservation
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:center;width: 2.3cm; ">
0.331
</td>
<td style="text-align:center;width: 2.3cm; ">
0.339
</td>
<td style="text-align:center;width: 2.3cm; ">
0.377
</td>
<td style="text-align:center;width: 2.3cm; ">
0.338
</td>
<td style="text-align:center;width: 2.3cm; ">
0.416
</td>
<td style="text-align:center;width: 2.3cm; ">
-0.015
</td>
<td style="text-align:center;width: 2.3cm; ">
0.010
</td>
<td style="text-align:center;width: 2.3cm; ">
-0.786
</td>
<td style="text-align:center;width: 2.3cm; ">
not_significant
</td>
</tr>
<tr>
<td style="text-align:center;width: 2.3cm; ">
0.336
</td>
<td style="text-align:center;width: 2.3cm; ">
0.342
</td>
<td style="text-align:center;width: 2.3cm; ">
0.380
</td>
<td style="text-align:center;width: 2.3cm; ">
0.338
</td>
<td style="text-align:center;width: 2.3cm; ">
0.422
</td>
<td style="text-align:center;width: 2.3cm; ">
0.046
</td>
<td style="text-align:center;width: 2.3cm; ">
0.009
</td>
<td style="text-align:center;width: 2.3cm; ">
2.194
</td>
<td style="text-align:center;width: 2.3cm; ">
diverged
</td>
</tr>
<tr>
<td style="text-align:center;width: 2.3cm; ">
0.300
</td>
<td style="text-align:center;width: 2.3cm; ">
0.308
</td>
<td style="text-align:center;width: 2.3cm; ">
0.343
</td>
<td style="text-align:center;width: 2.3cm; ">
0.311
</td>
<td style="text-align:center;width: 2.3cm; ">
0.376
</td>
<td style="text-align:center;width: 2.3cm; ">
-0.013
</td>
<td style="text-align:center;width: 2.3cm; ">
0.014
</td>
<td style="text-align:center;width: 2.3cm; ">
-0.800
</td>
<td style="text-align:center;width: 2.3cm; ">
not_significant
</td>
</tr>
<tr>
<td style="text-align:center;width: 2.3cm; ">
0.282
</td>
<td style="text-align:center;width: 2.3cm; ">
0.291
</td>
<td style="text-align:center;width: 2.3cm; ">
0.324
</td>
<td style="text-align:center;width: 2.3cm; ">
0.295
</td>
<td style="text-align:center;width: 2.3cm; ">
0.354
</td>
<td style="text-align:center;width: 2.3cm; ">
-0.005
</td>
<td style="text-align:center;width: 2.3cm; ">
0.018
</td>
<td style="text-align:center;width: 2.3cm; ">
-0.346
</td>
<td style="text-align:center;width: 2.3cm; ">
not_significant
</td>
</tr>
<tr>
<td style="text-align:center;width: 2.3cm; ">
0.311
</td>
<td style="text-align:center;width: 2.3cm; ">
0.317
</td>
<td style="text-align:center;width: 2.3cm; ">
0.354
</td>
<td style="text-align:center;width: 2.3cm; ">
0.324
</td>
<td style="text-align:center;width: 2.3cm; ">
0.383
</td>
<td style="text-align:center;width: 2.3cm; ">
-0.002
</td>
<td style="text-align:center;width: 2.3cm; ">
0.018
</td>
<td style="text-align:center;width: 2.3cm; ">
-0.148
</td>
<td style="text-align:center;width: 2.3cm; ">
not_significant
</td>
</tr>
<tr>
<td style="text-align:center;width: 2.3cm; ">
0.352
</td>
<td style="text-align:center;width: 2.3cm; ">
0.360
</td>
<td style="text-align:center;width: 2.3cm; ">
0.398
</td>
<td style="text-align:center;width: 2.3cm; ">
0.366
</td>
<td style="text-align:center;width: 2.3cm; ">
0.429
</td>
<td style="text-align:center;width: 2.3cm; ">
-0.008
</td>
<td style="text-align:center;width: 2.3cm; ">
0.016
</td>
<td style="text-align:center;width: 2.3cm; ">
-0.517
</td>
<td style="text-align:center;width: 2.3cm; ">
not_significant
</td>
</tr>
</tbody>
</table>
<p>In total, there are 57 modules that fulfill the minimum criteria of
being monophyletic for the human replicates:</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">module_conservation_human</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 57</span></span></code></pre></div>
<p>Among these, we identified 2 modules that are diverged on the human
lineage:</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">module_conservation_human</span><span class="op">$</span><span class="va">conservation</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;        diverged not_significant </span></span>
<span><span class="co">#&gt;               2              55</span></span></code></pre></div>
<p>The results can also be plotted using
<tt>plotConservedDivergedModules</tt>, the same function that has
already been used in case of the overall conservation and
divergence:</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotConservedDivergedModules.html">plotConservedDivergedModules</a></span><span class="op">(</span><span class="va">module_conservation_human</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="module_conservation_human.png" alt="Weighted linear regression between the human branch length and human diversity, with the diverged modules marked." width="75%"><p class="caption">
Weighted linear regression between the human branch length and human
diversity, with the diverged modules marked.
</p>
</div>
<p>The distance matrices and tree reconstructions of the ARID1B and MXI1
modules can showcase in more detail what makes them diverged on the
human lineage:</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">human_div_modules</span> <span class="op">&lt;-</span> <span class="va">module_conservation_human</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">conservation</span> <span class="op">==</span> <span class="st">"diverged"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">regulator</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dist</span> <span class="op">&lt;-</span> <span class="va">dist_jk</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">human_div_modules</span>, <span class="st">"_orig"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">dist</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">human_div_modules</span></span>
<span><span class="fu"><a href="../reference/plotDistMats.html">plotDistMats</a></span><span class="op">(</span><span class="va">dist</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="dist_mat_human_div.png" alt="The distance matrices of the 2 modules that were found to be diverged on the human lineage." width="70%"><p class="caption">
The distance matrices of the 2 modules that were found to be diverged on
the human lineage.
</p>
</div>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trees</span> <span class="op">&lt;-</span> <span class="va">trees_jk</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">human_div_modules</span>, <span class="st">"_orig"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">trees</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">human_div_modules</span></span>
<span><span class="fu"><a href="../reference/plotTrees.html">plotTrees</a></span><span class="op">(</span><span class="va">trees</span>, species_colors <span class="op">=</span> <span class="va">spec_colors</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="trees_human_div.png" alt="The tree reconstructions of the 2 modules that were found to be diverged on the human lineage." width="70%"><p class="caption">
The tree reconstructions of the 2 modules that were found to be diverged
on the human lineage.
</p>
</div>
<p>The signal - unsurprisingly - is not as strong as in case of the
modules diverged between cynomolgus and great apes, but it is still
apparent that the human-gorilla and human-cynomolgus distance measures
are the higher than the rest. This also translates into a way longer
human branch length than expected based on the phylogeny and observed in
most module trees.</p>
</div>
<div class="section level4">
<h4 id="finding-conserved-and-diverged-target-genes-1">Finding conserved and diverged target genes<a class="anchor" aria-label="anchor" href="#finding-conserved-and-diverged-target-genes-1"></a>
</h4>
<p>Similar to the approach in case of the overall conservation and
divergence, jackknifing can inform us about target genes that contribute
the most to human-specific divergence. In this case, the residuals of
the jackknifed modules versions are calculated with regard to the
regression model between the human branch length and the human
diversity. The target genes that result in the lowest residuals (lowest
signal is of divergence) when removed are considered the most diverged
in the original module.</p>
<p>We performed this analysis for the ARID1B module:</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ARID1B_target_conservation</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findConservedDivergedTargets.html">findConservedDivergedTargets</a></span><span class="op">(</span><span class="st">"ARID1B"</span>, </span>
<span>                                                           <span class="va">tree_stats_jk</span>, </span>
<span>                                                           <span class="va">lm_human</span><span class="op">)</span></span></code></pre></div>
<table class="table table" style="font-size: 13px; margin-left: auto; margin-right: auto;">
<caption style="font-size: initial !important;">
<span style="font-size:13px">Cross-species conservation measures for a
few example target genes of ARID1B</span>
</caption>
<thead><tr>
<th style="text-align:center;">
focus
</th>
<th style="text-align:center;">
regulator
</th>
<th style="text-align:center;">
type
</th>
<th style="text-align:center;">
id
</th>
<th style="text-align:center;">
gene_ removed
</th>
<th style="text-align:center;">
human_ diversity
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:center;width: 2.1cm; ">
human
</td>
<td style="text-align:center;width: 2.1cm; ">
ARID1B
</td>
<td style="text-align:center;width: 1.9cm; ">
orig
</td>
<td style="text-align:center;width: 3.1cm; ">
ARID1B_orig
</td>
<td style="text-align:center;width: 2.4cm; ">
NA
</td>
<td style="text-align:center;width: 2.2cm; ">
0.336
</td>
</tr>
<tr>
<td style="text-align:center;width: 2.1cm; ">
human
</td>
<td style="text-align:center;width: 2.1cm; ">
ARID1B
</td>
<td style="text-align:center;width: 1.9cm; ">
jk
</td>
<td style="text-align:center;width: 3.1cm; ">
ARID1B_jk_ACAT2
</td>
<td style="text-align:center;width: 2.4cm; ">
ACAT2
</td>
<td style="text-align:center;width: 2.2cm; ">
0.332
</td>
</tr>
<tr>
<td style="text-align:center;width: 2.1cm; ">
human
</td>
<td style="text-align:center;width: 2.1cm; ">
ARID1B
</td>
<td style="text-align:center;width: 1.9cm; ">
jk
</td>
<td style="text-align:center;width: 3.1cm; ">
ARID1B_jk_C12orf45
</td>
<td style="text-align:center;width: 2.4cm; ">
C12orf45
</td>
<td style="text-align:center;width: 2.2cm; ">
0.336
</td>
</tr>
<tr>
<td style="text-align:center;width: 2.1cm; ">
human
</td>
<td style="text-align:center;width: 2.1cm; ">
ARID1B
</td>
<td style="text-align:center;width: 1.9cm; ">
jk
</td>
<td style="text-align:center;width: 3.1cm; ">
ARID1B_jk_CHCHD2
</td>
<td style="text-align:center;width: 2.4cm; ">
CHCHD2
</td>
<td style="text-align:center;width: 2.2cm; ">
0.305
</td>
</tr>
</tbody>
</table>
<table class="table table" style="font-size: 13px; margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:center;">
human_ subtree_length
</th>
<th style="text-align:center;">
fit
</th>
<th style="text-align:center;">
lwr_fit
</th>
<th style="text-align:center;">
upr_fit
</th>
<th style="text-align:center;">
residual
</th>
<th style="text-align:center;">
contribution
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:center;width: 2.3cm; ">
0.428
</td>
<td style="text-align:center;width: 2.3cm; ">
0.378
</td>
<td style="text-align:center;width: 2.3cm; ">
0.335
</td>
<td style="text-align:center;width: 2.3cm; ">
0.420
</td>
<td style="text-align:center;width: 2.3cm; ">
0.050
</td>
<td style="text-align:center;width: 2.3cm; ">
NA
</td>
</tr>
<tr>
<td style="text-align:center;width: 2.3cm; ">
0.423
</td>
<td style="text-align:center;width: 2.3cm; ">
0.373
</td>
<td style="text-align:center;width: 2.3cm; ">
0.331
</td>
<td style="text-align:center;width: 2.3cm; ">
0.415
</td>
<td style="text-align:center;width: 2.3cm; ">
0.050
</td>
<td style="text-align:center;width: 2.3cm; ">
0.000
</td>
</tr>
<tr>
<td style="text-align:center;width: 2.3cm; ">
0.422
</td>
<td style="text-align:center;width: 2.3cm; ">
0.378
</td>
<td style="text-align:center;width: 2.3cm; ">
0.336
</td>
<td style="text-align:center;width: 2.3cm; ">
0.421
</td>
<td style="text-align:center;width: 2.3cm; ">
0.043
</td>
<td style="text-align:center;width: 2.3cm; ">
0.130
</td>
</tr>
<tr>
<td style="text-align:center;width: 2.3cm; ">
0.410
</td>
<td style="text-align:center;width: 2.3cm; ">
0.344
</td>
<td style="text-align:center;width: 2.3cm; ">
0.302
</td>
<td style="text-align:center;width: 2.3cm; ">
0.387
</td>
<td style="text-align:center;width: 2.3cm; ">
0.066
</td>
<td style="text-align:center;width: 2.3cm; ">
-0.312
</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotConservedDivergedTargets.html">plotConservedDivergedTargets</a></span><span class="op">(</span><span class="va">ARID1B_target_conservation</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="ARID1B_target_conservation.png" alt="The top 5 most diverged target genes of ARID1B identified using tree-based statistics in combination with jackknifing." width="75%"><p class="caption">
The top 5 most diverged target genes of ARID1B identified using
tree-based statistics in combination with jackknifing.
</p>
</div>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top5_diverged_ARID1B_targets</span> <span class="op">&lt;-</span> <span class="va">ARID1B_target_conservation</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">slice_min</span><span class="op">(</span>order_by <span class="op">=</span> <span class="va">residual</span>, n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">pull</span><span class="op">(</span><span class="va">gene_removed</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotExprAlongPseudotime.html">plotExprAlongPseudotime</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ARID1B"</span>, <span class="va">top5_diverged_ARID1B_targets</span><span class="op">)</span>, </span>
<span>                        <span class="va">sce</span>, </span>
<span>                        species_colors <span class="op">=</span> <span class="va">spec_colors</span>, </span>
<span>                        cell_type_colors <span class="op">=</span> <span class="va">ct_colors</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="ARID1B_target_expression.png" alt="The expression profiles of ARID1B and its top 5 most diverged target genes across pseudotime trajectory per species." width="60%"><p class="caption">
The expression profiles of ARID1B and its top 5 most diverged target
genes across pseudotime trajectory per species.
</p>
</div>
<p>Based on the effects of jackknifing, DLK1 proves to be by far the
most diverged target gene of ARID1B. This is also reflected in its
expression profile: it is upregulated much more in humans during the
intermediate stage of the neural differentiation than in gorilla or
cynomolgus.</p>
<p>Some other target genes of ARID1B also show slight human-specific
differences in their expression patterns, e.g. TMSB15A and DTX4 are both
upregulated the most in humans towards the late pseudotime stage.</p>
<p><br><br></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<p>The CroCoNet approach defines joint co-expression modules across all
species based on a phylogeny-aware consensus network and quantifies the
conservation of these modules by contrasting cross-species and
within-species topological differences. In addition, it can pinpoint
individual target genes that contribute the most to
conservation/divergence with the help of jackknifing. The modules and
target genes detected as conserved or diverged are good candidates to
explain cross-species differences and similarities in gene regulation
and can be deeper investigated by follow-up analyses and
experiments.</p>
<p><br><br></p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body">
<div id="ref-Shumate2021-ne" class="csl-entry">
1. Shumate A, Salzberg SL. <span class="nocase">Liftoff: accurate
mapping of gene annotations</span>. Bioinformatics. 2021;37:1639–43. <a href="https://doi.org/10.1093/bioinformatics/btaa1016" class="external-link">https://doi.org/10.1093/bioinformatics/btaa1016</a>.
</div>
<div id="ref-Rosen2024-li" class="csl-entry">
2. Rosen Y, Brbić M, Roohani Y, Swanson K, Li Z, Leskovec J. <span class="nocase">Toward universal cell embeddings: integrating single-cell
RNA-seq datasets across species with <span>SATURN</span></span>. Nat
Methods. 2024. <a href="https://doi.org/10.1038/s41592-024-02191-z" class="external-link">https://doi.org/10.1038/s41592-024-02191-z</a>.
</div>
<div id="ref-Lun2016-lx" class="csl-entry">
3. Lun ATL, McCarthy DJ, Marioni JC. <span class="nocase">A step-by-step
workflow for low-level analysis of single-cell RNA-seq data with
Bioconductor</span>. F1000Res. 2016;5:2122. <a href="https://doi.org/10.12688/f1000research.9501.2" class="external-link">https://doi.org/10.12688/f1000research.9501.2</a>.
</div>
<div id="ref-Cannoodt2016-hn" class="csl-entry">
4. Cannoodt R, Saelens W, Sichien D, Tavernier S, Janssens S, Guilliams
M, et al. <span class="nocase">SCORPIUS improves trajectory inference
and identifies novel modules in dendritic cell development</span>.
bioRxiv. 2016;079509. <a href="https://doi.org/10.1101/079509" class="external-link">https://doi.org/10.1101/079509</a>.
</div>
<div id="ref-Aran2019-jl" class="csl-entry">
5. Aran D, Looney AP, Liu L, Wu E, Fong V, Hsu A, et al. <span class="nocase">Reference-based analysis of lung single-cell sequencing
reveals a transitional profibrotic macrophage</span>. Nat Immunol.
2019;20:163–72. <a href="https://doi.org/10.1038/s41590-018-0276-y" class="external-link">https://doi.org/10.1038/s41590-018-0276-y</a>.
</div>
<div id="ref-Rhodes2022-os" class="csl-entry">
6. Rhodes K, Barr KA, Popp JM, Strober BJ, Battle A, Gilad Y. <span class="nocase">Human embryoid bodies as a novel system for genomic
studies of functionally diverse cell types</span>. Elife. 2022;11. <a href="https://doi.org/10.7554/eLife.71361" class="external-link">https://doi.org/10.7554/eLife.71361</a>.
</div>
<div id="ref-Moerman2019-xh" class="csl-entry">
7. Moerman T, Santos SA, González-Blas CB, Simm J, Moreau Y, Aerts J, et
al. <span class="nocase">GRNBoost2 and Arboreto: efficient and scalable
inference of gene regulatory networks</span>. Bioinformatics.
2019;35:2159–61. <a href="https://doi.org/10.1093/BIOINFORMATICS/BTY916" class="external-link">https://doi.org/10.1093/BIOINFORMATICS/BTY916</a>.
</div>
<div id="ref-Csardi2006-sy" class="csl-entry">
8. Csardi G, Nepusz T. <a href="https://igraph.org" class="external-link">The igraph software
package for complex network research</a>. InterJournal. 2006;Complex
Systems:1695.
</div>
<div id="ref-Agrawal2002-tb" class="csl-entry">
9. Agrawal H. <span class="nocase">Extreme Self-Organization in Networks
Constructed from Gene Expression Data</span>. Phys Rev Lett.
2002;89:268702. <a href="https://doi.org/10.1103/PHYSREVLETT.89.268702/FIGURES/3/MEDIUM" class="external-link">https://doi.org/10.1103/PHYSREVLETT.89.268702/FIGURES/3/MEDIUM</a>.
</div>
<div id="ref-Bergmann2003-sp" class="csl-entry">
10. Bergmann S, Ihmels J, Barkai N. <span class="nocase">Similarities
and Differences in Genome-Wide Expression Data of Six Organisms</span>.
PLoS Biol. 2003;2:e9. <a href="https://doi.org/10.1371/JOURNAL.PBIO.0020009" class="external-link">https://doi.org/10.1371/JOURNAL.PBIO.0020009</a>.
</div>
<div id="ref-Barabasi2004-gi" class="csl-entry">
11. Barabási AL, Oltvai ZN. <span class="nocase">Network biology:
understanding the cell’s functional organization</span>. Nature Reviews
Genetics 2004 5:2. 2004;5:101–13. <a href="https://doi.org/10.1038/nrg1272" class="external-link">https://doi.org/10.1038/nrg1272</a>.
</div>
<div id="ref-Broido2019-tf" class="csl-entry">
12. Broido AD, Clauset A. <span class="nocase">Scale-free networks are
rare</span>. Nature Communications 2019 10:1. 2019;10:1–10. <a href="https://doi.org/10.1038/s41467-019-08746-5" class="external-link">https://doi.org/10.1038/s41467-019-08746-5</a>.
</div>
<div id="ref-Bininda-Emonds2007-of" class="csl-entry">
13. Bininda-Emonds ORP, Cardillo M, Jones KE, MacPhee RDE, Beck RMD,
Grenyer R, et al. <span class="nocase">The delayed rise of present-day
mammals</span>. Nature. 2007;446:507–12. <a href="https://doi.org/10.1038/nature05634" class="external-link">https://doi.org/10.1038/nature05634</a>.
</div>
<div id="ref-Rauluseviciute2023-dz" class="csl-entry">
14. Rauluseviciute I, Riudavets-Puig R, Blanc-Mathieu R,
Castro-Mondragon JA, Ferenc K, Kumar V, et al. <span class="nocase">JASPAR 2024: 20th anniversary of the open-access database
of transcription factor binding profiles</span>. Nucleic Acids Res.
2023;52:D174–82. <a href="https://doi.org/10.1093/nar/gkad1059" class="external-link">https://doi.org/10.1093/nar/gkad1059</a>.
</div>
<div id="ref-Madsen2018-gq" class="csl-entry">
15. Madsen JGS, Rauch A, Van Hauwaert EL, Schmidt SF, Winnefeld M,
Mandrup S. <span class="nocase">Integrated analysis of motif activity
and gene expression changes of transcription factors</span>. Genome Res.
2018;28:243–55. <a href="https://doi.org/10.1101/gr.227231.117" class="external-link">https://doi.org/10.1101/gr.227231.117</a>.
</div>
<div id="ref-Langfelder2008-qe" class="csl-entry">
16. Langfelder P, Horvath S. <span class="nocase">WGCNA: An R package
for weighted correlation network analysis</span>. BMC Bioinformatics.
2008;9:1–13. <a href="https://doi.org/10.1186/1471-2105-9-559" class="external-link">https://doi.org/10.1186/1471-2105-9-559</a>.
</div>
<div id="ref-Ashburner2000-mc" class="csl-entry">
17. Ashburner M, Ball CA, Blake JA, Botstein D, Butler H, Cherry JM, et
al. <span class="nocase">Gene ontology: tool for the unification of
biology. The Gene Ontology Consortium</span>. Nat Genet. 2000;25:25–9.
<a href="https://doi.org/10.1038/75556" class="external-link">https://doi.org/10.1038/75556</a>.
</div>
<div id="ref-Gene-Ontology-Consortium2023-bs" class="csl-entry">
18. Gene Ontology Consortium, Aleksander SA, Balhoff J, Carbon S, Cherry
JM, Drabkin HJ, et al. <span class="nocase">The Gene Ontology
knowledgebase in <span>2023</span></span>. Genetics. 2023;224:iyad031.
<a href="https://doi.org/10.1093/genetics/iyad031" class="external-link">https://doi.org/10.1093/genetics/iyad031</a>.
</div>
<div id="ref-Alexa2024-oi" class="csl-entry">
19. Alexa A, Rahnenfuhrer J. <span class="nocase">topGO: enrichment
analysis for gene ontology</span>. R package version. 2024;2.58.0.
</div>
<div id="ref-Milacic2024-rx" class="csl-entry">
20. Milacic M, Beavers D, Conley P, Gong C, Gillespie M, Griss J, et al.
<span>The Reactome Pathway Knowledgebase <span>2024</span></span>.
Nucleic Acids Res. 2024;52:D672–8. <a href="https://doi.org/10.1093/nar/gkad1025" class="external-link">https://doi.org/10.1093/nar/gkad1025</a>.
</div>
<div id="ref-Yu2016-si" class="csl-entry">
21. Yu G, He Q-Y. <span class="nocase">ReactomePA: an R/Bioconductor
package for reactome pathway analysis and visualization</span>. Mol
Biosyst. 2016;12:477–9. <a href="https://doi.org/10.1039/c5mb00663e" class="external-link">https://doi.org/10.1039/c5mb00663e</a>.
</div>
<div id="ref-Szklarczyk2023-hx" class="csl-entry">
22. Szklarczyk D, Kirsch R, Koutrouli M, Nastou K, Mehryary F, Hachilif
R, et al. <span class="nocase">The STRING database in 2023:
protein-protein association networks and functional enrichment analyses
for any sequenced genome of interest</span>. Nucleic Acids Res.
2023;51:D638–46. <a href="https://doi.org/10.1093/nar/gkac1000" class="external-link">https://doi.org/10.1093/nar/gkac1000</a>.
</div>
<div id="ref-Yu2012-bv" class="csl-entry">
23. Yu G, Wang L-G, Han Y, He Q-Y. <span class="nocase">clusterProfiler:
an R package for comparing biological themes among gene clusters</span>.
OMICS. 2012;16:284–7. <a href="https://doi.org/10.1089/omi.2011.0118" class="external-link">https://doi.org/10.1089/omi.2011.0118</a>.
</div>
<div id="ref-Langfelder2011-wm" class="csl-entry">
24. Langfelder P, Luo R, Oldham MC, Horvath S. <span class="nocase">Is
my network module preserved and reproducible?</span> PLoS Comput Biol.
2011;7:1001057. <a href="https://doi.org/10.1371/journal.pcbi.1001057" class="external-link">https://doi.org/10.1371/journal.pcbi.1001057</a>.
</div>
<div id="ref-Wang2014-ew" class="csl-entry">
25. Wang J, Xie G, Singh M, Ghanbarian AT, Raskó T, Szvetnik A, et al.
<span class="nocase">Primate-specific endogenous retrovirus-driven
transcription defines naive-like stem cells</span>. Nature.
2014;516:405–9. <a href="https://doi.org/10.1038/nature13804" class="external-link">https://doi.org/10.1038/nature13804</a>.
</div>
<div id="ref-Hsieh2022-yl" class="csl-entry">
26. Hsieh F-K, Ji F, Damle M, Sadreyev RI, Kingston RE. <span class="nocase">HERVH-derived lncRNAs negatively regulate chromatin
targeting and remodeling mediated by <span>CHD7</span></span>. Life Sci
Alliance. 2022;5:e202101127. <a href="https://doi.org/10.26508/lsa.202101127" class="external-link">https://doi.org/10.26508/lsa.202101127</a>.
</div>
<div id="ref-Ito2017-da" class="csl-entry">
27. Ito J, Sugimoto R, Nakaoka H, Yamada S, Kimura T, Hayano T, et al.
<span class="nocase">Systematic identification and characterization of
regulatory elements derived from human endogenous retroviruses</span>.
PLoS Genet. 2017;13:e1006883. <a href="https://doi.org/10.1371/journal.pgen.1006883" class="external-link">https://doi.org/10.1371/journal.pgen.1006883</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Anita Térmeg.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
